// spectrum-ts/dist/chunk-5XW4CAWS.js
// src/content/attachment.ts
import { createReadStream } from "fs";
import { readFile, stat } from "fs/promises";
import { basename } from "path";
import { Readable } from "stream";
import { lookup as lookupMimeType } from "mime-types";
import z from "zod";
var DEFAULT_ATTACHMENT_NAME = "attachment";
var readSchema = z.function({
  input: [],
  output: z.promise(z.instanceof(Buffer))
});
var streamSchema = z.function({
  input: [],
  output: z.promise(z.instanceof(ReadableStream))
});
var attachmentSchema = z.object({
  type: z.literal("attachment"),
  name: z.string().nonempty(),
  mimeType: z.string().nonempty(),
  size: z.number().int().nonnegative().optional(),
  read: readSchema,
  stream: streamSchema
});
var resolveAttachmentName = (input, name) => name || (typeof input === "string" ? basename(input) : DEFAULT_ATTACHMENT_NAME);
var resolveAttachmentMimeType = (name, mimeType) => {
  if (mimeType) {
    return mimeType;
  }
  const resolvedMimeType = lookupMimeType(name);
  if (!resolvedMimeType) {
    throw new Error(
      `Unable to resolve MIME type for attachment "${name}". Pass options.mimeType explicitly.`
    );
  }
  return resolvedMimeType;
};
var bufferToStream = (buf) => new ReadableStream({
  start(controller) {
    controller.enqueue(buf);
    controller.close();
  }
});
var asAttachment = (input) => {
  let cached;
  const read = () => {
    cached ??= input.read().catch((err) => {
      cached = void 0;
      throw err;
    });
    return cached;
  };
  const stream2 = input.stream ?? (async () => bufferToStream(await read()));
  return attachmentSchema.parse({
    type: "attachment",
    name: input.name,
    mimeType: input.mimeType,
    size: input.size,
    read,
    stream: stream2
  });
};
function attachment(input, options) {
  return {
    build: async () => {
      const name = resolveAttachmentName(input, options?.name);
      const mimeType = resolveAttachmentMimeType(name, options?.mimeType);
      if (typeof input === "string") {
        const stats = await stat(input);
        return asAttachment({
          name,
          mimeType,
          size: stats.size,
          read: () => readFile(input),
          stream: async () => Readable.toWeb(
            createReadStream(input)
          )
        });
      }
      return asAttachment({
        name,
        mimeType,
        size: input.byteLength,
        read: async () => input,
        stream: async () => bufferToStream(input)
      });
    }
  };
}

// src/content/custom.ts
import z2 from "zod";
var customSchema = z2.object({
  type: z2.literal("custom"),
  raw: z2.unknown()
});
var asCustom = (raw) => customSchema.parse({ type: "custom", raw });
function custom(raw) {
  return {
    build: async () => asCustom(raw)
  };
}

// src/utils/stream.ts
import { Repeater } from "@repeaterjs/repeater";
function stream(setup) {
  const repeater = new Repeater(async (push, stop) => {
    const emit = (value) => {
      Promise.resolve(push(value)).catch((error) => {
        stop(error);
        return void 0;
      });
    };
    const end = (error) => {
      stop(error);
    };
    const cleanup = await setup(emit, end);
    try {
      await stop;
    } finally {
      await cleanup?.();
    }
  });
  return Object.assign(repeater, {
    close: async () => {
      await repeater.return(void 0);
    }
  });
}
function mergeStreams(streams) {
  return stream((emit, end) => {
    if (streams.length === 0) {
      end();
      return;
    }
    let openStreams = streams.length;
    const workers = streams.map(async (source) => {
      try {
        for await (const value of source) {
          emit(value);
        }
      } catch (error) {
        end(error);
      } finally {
        openStreams -= 1;
        if (openStreams === 0) {
          end();
        }
      }
    });
    return async () => {
      await Promise.allSettled(streams.map((source) => source.close()));
      await Promise.allSettled(workers);
    };
  });
}

export {
  asAttachment,
  attachment,
  asCustom,
  custom,
  stream,
  mergeStreams
};
