// spectrum-ts/dist/chunk-HXM64ENV.js
// src/utils/cloud.ts
var SPECTRUM_CLOUD_URL = `https://${process.env.SPECTRUM_CLOUD_URL ?? "spectrum.photon.codes"}`;
var SpectrumCloudError = class extends Error {
  status;
  code;
  constructor(status, code, message) {
    super(message);
    this.name = "SpectrumCloudError";
    this.status = status;
    this.code = code;
  }
};
var request = async (path, init) => {
  const response = await fetch(`${SPECTRUM_CLOUD_URL}${path}`, init);
  if (!response.ok) {
    const body = await response.text().catch(() => "");
    try {
      const parsed = JSON.parse(body);
      throw new SpectrumCloudError(
        response.status,
        parsed.code,
        parsed.message
      );
    } catch (error) {
      if (error instanceof SpectrumCloudError) {
        throw error;
      }
      throw new SpectrumCloudError(
        response.status,
        "UNKNOWN",
        body || response.statusText
      );
    }
  }
  const json = await response.json();
  if (!json.succeed) {
    throw new SpectrumCloudError(
      response.status,
      "UNKNOWN",
      "Server returned succeed=false"
    );
  }
  return json.data;
};
var basicAuth = (projectId, projectSecret) => `Basic ${btoa(`${projectId}:${projectSecret}`)}`;
var cloud = {
  getSubscription: (projectId) => request(`/projects/${projectId}/billing/subscription`),
  issueImessageTokens: (projectId, projectSecret) => request(`/projects/${projectId}/imessage/tokens`, {
    method: "POST",
    headers: { Authorization: basicAuth(projectId, projectSecret) }
  }),
  getImessageInfo: (projectId) => request(`/projects/${projectId}/imessage/`),
  getPlatforms: (projectId) => request(`/projects/${projectId}/platforms/`),
  togglePlatform: (projectId, projectSecret, platform, enabled) => request(`/projects/${projectId}/platforms/`, {
    method: "PATCH",
    headers: {
      Authorization: basicAuth(projectId, projectSecret),
      "Content-Type": "application/json"
    },
    body: JSON.stringify({ platform, enabled })
  })
};

export {
  SpectrumCloudError,
  cloud
};
