// spectrum-ts/dist/index.js
import {
  SpectrumCloudError,
  cloud
} from "./chunk-HXM64ENV.js";
import {
  attachment,
  custom,
  mergeStreams,
  stream
} from "./chunk-5XW4CAWS.js";
import {
  definePlatform,
  resolveContents,
  text
} from "./chunk-XEEDIGVK.js";

// src/spectrum.ts
import z from "zod";
var providerMessageCoreKeys = /* @__PURE__ */ new Set([
  "content",
  "id",
  "sender",
  "space",
  "timestamp"
]);
var spectrumConfigSchema = z.union([
  z.object({
    projectId: z.string().min(1),
    projectSecret: z.string().min(1),
    providers: z.array(z.custom())
  }),
  z.object({
    projectId: z.undefined().optional(),
    projectSecret: z.undefined().optional(),
    providers: z.array(z.custom())
  })
]);
async function Spectrum(options) {
  spectrumConfigSchema.parse(options);
  const { projectId, projectSecret, providers } = options;
  const platformStates = /* @__PURE__ */ new Map();
  const customEventStreams = /* @__PURE__ */ new Map();
  let stopped = false;
  for (const provider of providers) {
    const providerConfig = provider;
    const def = providerConfig.__definition;
    const userConfig = def.config.parse(providerConfig.config);
    const client = await def.lifecycle.createClient({
      config: userConfig,
      projectId,
      projectSecret
    });
    platformStates.set(def.name, {
      client,
      config: userConfig,
      definition: def
    });
  }
  const adaptIterable = (iterable) => {
    return stream((emit, end) => {
      const iterator = iterable[Symbol.asyncIterator]();
      (async () => {
        try {
          let result = await iterator.next();
          while (!result.done) {
            emit(result.value);
            result = await iterator.next();
          }
          end();
        } catch (error) {
          end(error);
        }
      })();
      return async () => {
        await iterator.return?.();
      };
    });
  };
  const createProviderMessagesStream = (state) => {
    const { client, config, definition } = state;
    const raw = definition.events.messages({
      client,
      config
    });
    const bindSend = async function* () {
      for await (const msg of raw) {
        const extraEntries = Object.entries(msg).filter(
          ([key]) => !providerMessageCoreKeys.has(key)
        );
        const extra = Object.fromEntries(extraEntries);
        const parsedExtra = definition.message?.schema ? definition.message.schema.parse(extra) : {};
        const spaceRef = {
          ...msg.space,
          __platform: definition.name
        };
        const typingCtx = { space: spaceRef, client, config };
        const space = {
          ...spaceRef,
          send: async (...content) => {
            const resolved = await resolveContents(content);
            for (const item of resolved) {
              await definition.actions.send({
                ...typingCtx,
                content: item
              });
            }
          },
          startTyping: async () => {
            await definition.actions.startTyping?.(typingCtx);
          },
          stopTyping: async () => {
            await definition.actions.stopTyping?.(typingCtx);
          },
          responding: async (fn) => {
            await definition.actions.startTyping?.(typingCtx);
            try {
              return await fn();
            } finally {
              await definition.actions.stopTyping?.(typingCtx).catch(() => {
              });
            }
          }
        };
        const normalizedMessage = {
          ...parsedExtra,
          id: msg.id,
          content: msg.content,
          platform: definition.name,
          react: async (reaction) => {
            if (!definition.actions.reactToMessage) {
              return;
            }
            await definition.actions.reactToMessage({
              space: spaceRef,
              messageId: msg.id,
              reaction,
              client,
              config
            });
          },
          reply: async (...content) => {
            if (!definition.actions.replyToMessage) {
              return;
            }
            const resolved = await resolveContents(content);
            for (const item of resolved) {
              await definition.actions.replyToMessage({
                space: spaceRef,
                messageId: msg.id,
                content: item,
                client,
                config
              });
            }
          },
          sender: {
            ...msg.sender,
            __platform: definition.name
          },
          space,
          timestamp: msg.timestamp ?? /* @__PURE__ */ new Date()
        };
        yield [space, normalizedMessage];
      }
    };
    return adaptIterable(bindSend());
  };
  const createMessagesStream = () => {
    return stream(async (emit, end) => {
      const merged = mergeStreams(
        Array.from(platformStates.values(), createProviderMessagesStream)
      );
      (async () => {
        try {
          for await (const value of merged) {
            emit(value);
          }
          end();
        } catch (error) {
          end(error);
        }
      })();
      return async () => {
        await merged.close();
      };
    });
  };
  const createCustomEventStream = (eventName) => {
    return stream(async (emit, end) => {
      const providerStreams = Array.from(platformStates.values(), (state) => {
        const { client, config, definition } = state;
        const producer = definition.events[eventName];
        if (!producer) {
          return void 0;
        }
        const providerEvents = producer({ client, config });
        const annotatePlatform = async function* () {
          for await (const value of providerEvents) {
            yield { ...value, platform: definition.name };
          }
        };
        return adaptIterable(annotatePlatform());
      }).filter(
        (value) => value !== void 0
      );
      const merged = mergeStreams(providerStreams);
      (async () => {
        try {
          for await (const value of merged) {
            emit(value);
          }
          end();
        } catch (error) {
          end(error);
        }
      })();
      return async () => {
        await merged.close();
      };
    });
  };
  const messagesStream = createMessagesStream();
  const stopOnce = async () => {
    if (stopped) {
      return;
    }
    stopped = true;
    const streamShutdowns = [
      messagesStream.close(),
      ...Array.from(
        customEventStreams.values(),
        (eventStream) => eventStream.close()
      )
    ];
    process.off("SIGINT", handleSignal);
    process.off("SIGTERM", handleSignal);
    await Promise.allSettled(streamShutdowns);
    const clientShutdowns = Array.from(
      platformStates.values(),
      (state) => state.definition.lifecycle.destroyClient({
        client: state.client
      })
    );
    await Promise.allSettled(clientShutdowns);
    customEventStreams.clear();
    platformStates.clear();
  };
  const handleSignal = () => {
    setTimeout(() => process.exit(1), 3e3).unref();
    stopOnce().then(() => process.exit(0)).catch(() => process.exit(1));
  };
  process.on("SIGINT", handleSignal);
  process.on("SIGTERM", handleSignal);
  const messages = messagesStream;
  const customEventProxy = new Proxy(
    {},
    {
      get(_target, prop) {
        let eventStream = customEventStreams.get(prop);
        if (!eventStream) {
          eventStream = createCustomEventStream(prop);
          customEventStreams.set(prop, eventStream);
        }
        return eventStream;
      }
    }
  );
  const base = {
    __providers: providers,
    __internal: { platforms: platformStates },
    messages,
    stop: stopOnce,
    send: async (space, ...content) => {
      await space.send(...content);
    },
    responding: async (space, fn) => {
      return space.responding(fn);
    }
  };
  return new Proxy(base, {
    get(target, prop, receiver) {
      if (prop in target) {
        return Reflect.get(target, prop, receiver);
      }
      if (typeof prop === "string") {
        return customEventProxy[prop];
      }
      return void 0;
    }
  });
}
export {
  Spectrum,
  SpectrumCloudError,
  attachment,
  cloud,
  custom,
  definePlatform,
  mergeStreams,
  resolveContents,
  stream,
  text
};
