// @photon-ai/imessage-kit/dist/index.cjs
'use strict';

var os = require('os');
var path = require('path');
var typedstream = require('@parseaple/typedstream');
var module$1 = require('module');
var fs = require('fs');
var child_process = require('child_process');
var util = require('util');
var crypto$1 = require('crypto');
var promises = require('fs/promises');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
// src/application/message-chain.ts
var MessageChain = class {
  constructor(message, sender) {
    this.message = message;
    this.sender = sender;
  }
  active = true;
  actions = [];
  executed = false;
  // -----------------------------------------------
  // Predicates
  // -----------------------------------------------
  /** Custom predicate. Short-circuits the chain if false. */
  when(predicate) {
    this.assertMutable();
    if (this.active) {
      this.active = predicate(this.message);
    }
    return this;
  }
  /** Text contains the string or matches the regex. */
  matchText(pattern) {
    return this.when((m) => {
      if (!m.text) return false;
      return typeof pattern === "string" ? m.text.includes(pattern) : pattern.test(m.text);
    });
  }
  /** Message is from another participant (not the local user). */
  ifFromOthers() {
    return this.when((m) => !m.isFromMe);
  }
  /** Message is from the local user. */
  ifFromMe() {
    return this.when((m) => m.isFromMe);
  }
  /** Message is in a group chat. */
  ifGroup() {
    return this.when((m) => m.chatKind === "group");
  }
  /** Message is a DM (1-on-1 chat). */
  ifDM() {
    return this.when((m) => m.chatKind === "dm");
  }
  /** Message is a reaction (tapback, sticker, poll vote). */
  ifReaction() {
    return this.when((m) => m.reaction !== null);
  }
  /** Message is not a reaction. */
  ifNotReaction() {
    return this.when((m) => m.reaction === null);
  }
  /** Message has not been read. */
  ifUnread() {
    return this.when((m) => !m.isRead);
  }
  // -----------------------------------------------
  // Actions
  // -----------------------------------------------
  /** Queue a text reply to the message's chat. */
  replyText(text) {
    this.assertMutable();
    if (this.active) {
      this.actions.push(async () => {
        const content = typeof text === "function" ? text(this.message) : text;
        await this.sender.send({ to: this.message.chatId, text: content });
      });
    }
    return this;
  }
  /** Queue an attachment reply to the message's chat. */
  replyAttachments(paths) {
    this.assertMutable();
    if (this.active) {
      this.actions.push(async () => {
        const selected = typeof paths === "function" ? paths(this.message) : paths;
        const attachments = Array.isArray(selected) ? selected : [selected];
        await this.sender.send({ to: this.message.chatId, attachments });
      });
    }
    return this;
  }
  /** Queue a custom handler. */
  do(handler) {
    this.assertMutable();
    if (this.active) {
      this.actions.push(async () => {
        await handler(this.message);
      });
    }
    return this;
  }
  // -----------------------------------------------
  // Execution
  // -----------------------------------------------
  /**
   * Execute all queued actions sequentially.
   *
   * Continues on error — all errors are collected in ChainResult.
   * Calling execute() more than once is a no-op.
   */
  async execute() {
    if (this.executed) return { actionsRun: 0, errors: [] };
    this.executed = true;
    if (!this.active || this.actions.length === 0) {
      return { actionsRun: 0, errors: [] };
    }
    const errors = [];
    let actionsRun = 0;
    for (const action of this.actions) {
      actionsRun++;
      try {
        await action();
      } catch (err) {
        errors.push(err instanceof Error ? err : new Error(String(err)));
      }
    }
    return { actionsRun, errors };
  }
  // -----------------------------------------------
  // Internal
  // -----------------------------------------------
  assertMutable() {
    if (this.executed) {
      throw new Error("MessageChain has already been executed");
    }
  }
};

// src/domain/errors.ts
var IMessageError = class _IMessageError extends Error {
  code;
  constructor(code, message, options) {
    super(message, options);
    this.code = code;
    this.name = "IMessageError";
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, _IMessageError);
    }
  }
};
function PlatformError(message = "Only macOS is supported", cause) {
  return new IMessageError("PLATFORM", message, { cause });
}
function DatabaseError(message, cause) {
  return new IMessageError("DATABASE", message, { cause });
}
function SendError(message, cause) {
  return new IMessageError("SEND", message, { cause });
}
function ConfigError(message, cause) {
  return new IMessageError("CONFIG", message, { cause });
}
function toError(value) {
  return value instanceof Error ? value : new Error(String(value));
}
function toErrorMessage(value) {
  return value instanceof Error ? value.message : String(value);
}

// src/application/message-dispatcher.ts
var MessageDispatcher = class {
  events;
  sink;
  outgoingMatcher;
  debug;
  dispatching = false;
  constructor(options = {}) {
    this.events = options.events ?? {};
    this.sink = options.sink;
    this.outgoingMatcher = options.outgoingMatcher;
    this.debug = options.debug ?? false;
  }
  /**
   * Process a batch of messages from the watch source.
   *
   * Pipeline: reconcile outgoing → filter isFromMe → dispatch to sink + events.
   *
   * Must not be called concurrently. The watcher must await the previous
   * dispatch before calling again.
   */
  async dispatch(messages) {
    if (this.dispatching) {
      throw new Error("dispatch() must not be called concurrently");
    }
    this.dispatching = true;
    try {
      this.reconcileOutgoing(messages);
      const incoming = messages.filter((m) => !m.isFromMe);
      await this.dispatchAll(incoming);
      this.outgoingMatcher?.cleanup();
    } finally {
      this.dispatching = false;
    }
  }
  /** Forward an external error to the sink and user callback. */
  handleError(error, context = "dispatcher") {
    const err = toError(error);
    if (this.debug) {
      console.error(`[MessageDispatcher] ${context}:`, err);
    }
    this.sink?.onError(err, context);
    this.events.onError?.(err);
  }
  // -----------------------------------------------
  // Internal
  // -----------------------------------------------
  reconcileOutgoing(messages) {
    if (!this.outgoingMatcher) return;
    for (const message of messages) {
      if (!message.isFromMe) continue;
      this.outgoingMatcher.tryMatch(message);
    }
  }
  async dispatchAll(messages) {
    for (const message of messages) {
      try {
        await this.dispatchOne(message);
      } catch (error) {
        this.handleError(error, "dispatch-message");
      }
    }
  }
  async dispatchOne(message) {
    await this.sink?.onMessage(message);
    await this.events.onMessage?.(message);
    switch (message.chatKind) {
      case "group":
        await this.events.onGroupMessage?.(message);
        break;
      case "dm":
        await this.events.onDirectMessage?.(message);
        break;
    }
  }
};

// src/domain/chat.ts
var CHAT_STYLE_GROUP = 43;
var CHAT_STYLE_DM = 45;
function resolveChatKind(style) {
  switch (style) {
    case CHAT_STYLE_GROUP:
      return "group";
    case CHAT_STYLE_DM:
      return "dm";
    default:
      return "unknown";
  }
}

// src/domain/timestamp.ts
var MAC_EPOCH = (/* @__PURE__ */ new Date("2001-01-01T00:00:00Z")).getTime();
function toMacTimestampNs(date) {
  return (date.getTime() - MAC_EPOCH) * 1e6;
}
function fromMacTimestampNs(ns) {
  return new Date(MAC_EPOCH + ns / 1e6);
}

// src/domain/chat-id.ts
var CHAT_SERVICE_PREFIX_PATTERN = /^[A-Za-z][A-Za-z0-9._-]*$/;
var GUID_MIN_LENGTH = 8;
var BARE_GUID_MIN_LENGTH = 10;
var BARE_GUID_PREFIX = "chat";
var GROUP_SEPARATOR = ";+;";
var DM_SEPARATOR = ";-;";
function parseChatServicePrefix(value) {
  if (value == null || value === "") return null;
  return CHAT_SERVICE_PREFIX_PATTERN.test(value) ? value : null;
}
function isValidPrefixedFormat(raw, separator) {
  const index = raw.indexOf(separator);
  if (index === -1) {
    return false;
  }
  const prefix = raw.slice(0, index);
  const suffix = raw.slice(index + separator.length);
  return parseChatServicePrefix(prefix) != null && suffix !== "";
}
var ChatId = class _ChatId {
  /** The raw chat identifier string. */
  raw;
  /** Whether this identifies a group conversation. */
  isGroup;
  constructor(raw, isGroup) {
    this.raw = raw;
    this.isGroup = isGroup;
  }
  // -----------------------------------------------
  // Factory methods
  // -----------------------------------------------
  /** Construct from a raw identifier string. */
  static fromUserInput(raw) {
    return new _ChatId(raw, _ChatId.detectGroup(raw));
  }
  /** Construct a DM chat id from a recipient address and optional service prefix. */
  static fromDMRecipient(recipient, prefix = "iMessage") {
    return new _ChatId(`${prefix}${DM_SEPARATOR}${recipient}`, false);
  }
  // -----------------------------------------------
  // Computed properties
  // -----------------------------------------------
  /** Chat kind as a string. */
  get kind() {
    return this.isGroup ? "group" : "dm";
  }
  /**
   * Core identifier with all service prefixes stripped.
   *
   * `iMessage;-;pilot@photon.codes` → `pilot@photon.codes`
   * `iMessage;+;chat613ABF`         → `chat613ABF`
   * `chat613ABF`                    → `chat613ABF`
   */
  get coreIdentifier() {
    return this.extractAfter(GROUP_SEPARATOR) ?? this.extractAfter(DM_SEPARATOR) ?? this.raw;
  }
  /**
   * Group identifier: the GUID portion for groups, raw for non-groups.
   */
  get groupIdentifier() {
    if (!this.isGroup) return this.raw;
    return this.extractAfter(GROUP_SEPARATOR) ?? this.raw;
  }
  // -----------------------------------------------
  // Methods
  // -----------------------------------------------
  /**
   * Extract the recipient address from a DM chat id.
   *
   * Returns `null` for group chats or bare values without a separator.
   */
  extractRecipient() {
    if (this.isGroup) return null;
    return this.extractAfter(DM_SEPARATOR);
  }
  /** Extract everything after a separator, or null if the separator is absent. */
  extractAfter(separator) {
    if (!this.raw.includes(separator)) return null;
    return this.raw.split(separator).slice(1).join(separator);
  }
  /**
   * Check whether this ChatId matches another.
   *
   * Compares raw strings and core identifiers (case-sensitive).
   */
  matches(other) {
    return this.raw === other.raw || this.coreIdentifier === other.coreIdentifier;
  }
  /**
   * Validate the chat id format.
   *
   * Throws `Error` for malformed inputs. Accepts the three documented formats:
   * `service;+;guid`, `service;-;address`, and bare GUIDs (≥8 chars, no semicolons).
   */
  validate() {
    if (this.raw === "") {
      throw new Error("ChatId cannot be empty");
    }
    if (!this.raw.includes(";")) {
      if (this.raw.length < GUID_MIN_LENGTH) {
        throw new Error(`Bare GUID too short: "${this.raw}" (minimum ${GUID_MIN_LENGTH} characters)`);
      }
      return;
    }
    if (isValidPrefixedFormat(this.raw, GROUP_SEPARATOR) || isValidPrefixedFormat(this.raw, DM_SEPARATOR)) {
      return;
    }
    throw new Error(`Malformed chat id: "${this.raw}" (expected service;+;guid or service;-;address)`);
  }
  /**
   * Build a full Messages.app guid with a service prefix.
   *
   * `chat613ABF` + `iMessage` → `iMessage;+;chat613ABF`
   */
  buildGroupGuid(prefix) {
    const core = this.groupIdentifier;
    return `${prefix}${GROUP_SEPARATOR}${core}`;
  }
  toString() {
    return this.raw;
  }
  // -----------------------------------------------
  // Internal
  // -----------------------------------------------
  static detectGroup(raw) {
    if (raw.includes(GROUP_SEPARATOR)) return true;
    if (!raw.includes(";") && raw.startsWith(BARE_GUID_PREFIX) && raw.length > BARE_GUID_MIN_LENGTH) {
      return true;
    }
    return false;
  }
};

// src/infra/db/contract.ts
function buildChatIdMatchSql(userInput, columns) {
  const core = ChatId.fromUserInput(userInput).coreIdentifier;
  if (core === userInput) {
    return {
      sql: `(${columns.identifier} = ? OR ${columns.guid} = ?)`,
      params: [userInput, userInput]
    };
  }
  return {
    sql: `(${columns.identifier} = ? OR ${columns.guid} = ? OR ${columns.identifier} = ?)`,
    params: [userInput, userInput, core]
  };
}

// src/infra/db/macos26.ts
var MESSAGE_FIELDS = [
  "message.ROWID as id",
  "message.guid",
  "message.text",
  "message.attributedBody",
  "message.service",
  "message.is_from_me",
  "message.is_read",
  "message.is_sent",
  "message.is_delivered",
  "message.was_downgraded",
  "message.did_notify_recipient",
  "message.is_auto_reply",
  "message.is_system_message",
  "message.is_forward",
  "message.is_audio_message",
  "message.is_played",
  "message.is_expirable",
  "message.error",
  "message.is_spam",
  "message.is_kt_verified",
  "message.has_unseen_mention",
  "message.was_delivered_quietly",
  "message.is_sos",
  "message.is_critical",
  "message.sent_or_received_off_grid",
  "message.date",
  "message.date_delivered",
  "message.date_read",
  "message.date_played",
  "message.date_edited",
  "message.date_retracted",
  "message.date_recovered",
  "message.reply_to_guid",
  "message.thread_originator_guid",
  "message.group_title",
  "message.expressive_send_style_id",
  "message.balloon_bundle_id",
  "message.destination_caller_id",
  "message.ck_chat_id as ck_chat_id",
  "message.was_detonated",
  "message.expire_state",
  "message.share_status",
  "message.share_direction",
  "message.schedule_type",
  "message.schedule_state",
  "message.part_count",
  "message.associated_message_type",
  "message.associated_message_guid",
  "message.associated_message_emoji",
  "message.associated_message_range_location",
  "message.associated_message_range_length",
  "message.item_type",
  "message.group_action_type",
  "handle.id as participant",
  "other_handle.id as affected_participant",
  "chat.chat_identifier as chat_id",
  "chat.guid as chat_guid",
  "chat.service_name as chat_service"
];
var CHAT_FIELDS = [
  "chat.guid",
  "chat.chat_identifier",
  "chat.service_name",
  "chat.style",
  "chat.account_login",
  "chat.is_archived",
  "chat.is_filtered",
  "chat.is_blackholed",
  "chat.is_deleting_incoming_messages",
  "chat.last_read_message_timestamp",
  "chat.display_name",
  "chat_stats.last_date",
  "COALESCE(chat_stats.unread_count, 0) AS unread_count"
];
var ATTACHMENT_FIELDS = [
  "message_attachment_join.message_id as msg_id",
  "attachment.guid",
  "attachment.created_date",
  "attachment.filename",
  "attachment.uti",
  "attachment.mime_type",
  "attachment.transfer_state",
  "attachment.is_outgoing",
  "attachment.transfer_name",
  "attachment.total_bytes",
  "attachment.is_sticker",
  "attachment.is_commsafety_sensitive",
  "attachment.emoji_image_short_description"
];
function escapeLikePattern(input) {
  return input.replace(/[%_\\]/g, (ch) => `\\${ch}`);
}
var macos26Queries = {
  schemaId: "macos26",
  buildMessageQuery(filter) {
    const conditions = [];
    const params = [];
    if (filter.unreadOnly) {
      conditions.push("message.is_read = 0");
    }
    if (filter.isFromMe === true) {
      conditions.push("message.is_from_me = 1");
    } else if (filter.isFromMe === false) {
      conditions.push("message.is_from_me = 0");
    }
    if (filter.participant) {
      conditions.push("handle.id = ?");
      params.push(filter.participant);
    }
    if (filter.chatId) {
      const match = buildChatIdMatchSql(filter.chatId, {
        identifier: "chat.chat_identifier",
        guid: "chat.guid"
      });
      conditions.push(match.sql);
      params.push(...match.params);
    }
    if (filter.service) {
      conditions.push("message.service = ?");
      params.push(filter.service);
    }
    if (filter.hasAttachments) {
      conditions.push(
        "EXISTS (SELECT 1 FROM message_attachment_join WHERE message_attachment_join.message_id = message.ROWID)"
      );
    }
    if (filter.excludeReactions) {
      conditions.push("(message.associated_message_type IS NULL OR message.associated_message_type = 0)");
    }
    if (filter.sinceRowId != null) {
      conditions.push("message.ROWID > ?");
      params.push(filter.sinceRowId);
    }
    if (filter.since) {
      conditions.push("message.date >= ?");
      params.push(toMacTimestampNs(filter.since));
    }
    if (filter.before) {
      conditions.push("message.date < ?");
      params.push(toMacTimestampNs(filter.before));
    }
    if (filter.search) {
      const escaped = escapeLikePattern(filter.search);
      conditions.push("message.text LIKE ? ESCAPE '\\'");
      params.push(`%${escaped}%`);
    }
    const where = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
    const hasLimit = filter.limit != null && filter.limit > 0;
    const hasOffset = filter.offset != null && filter.offset > 0;
    let limitClause = "";
    if (hasLimit) {
      limitClause = "LIMIT ?";
      params.push(filter.limit);
    } else if (hasOffset) {
      limitClause = "LIMIT -1";
    }
    const offsetClause = hasOffset ? "OFFSET ?" : "";
    if (hasOffset) {
      params.push(filter.offset);
    }
    const orderBy = filter.orderByRowIdAsc ? "ORDER BY message.ROWID ASC" : "ORDER BY message.date DESC";
    return {
      sql: `
                SELECT
                    ${MESSAGE_FIELDS.join(",\n                    ")}
                FROM message
                LEFT JOIN handle ON message.handle_id = handle.ROWID
                LEFT JOIN handle AS other_handle ON message.other_handle = other_handle.ROWID
                LEFT JOIN chat ON chat.ROWID = (
                    SELECT MIN(chat_message_join.chat_id)
                    FROM chat_message_join
                    WHERE chat_message_join.message_id = message.ROWID
                )
                ${where}
                ${orderBy}
                ${limitClause}
                ${offsetClause}
            `,
      params
    };
  },
  buildChatQuery(query) {
    const conditions = [];
    const params = [];
    if (query.chatId) {
      const match = buildChatIdMatchSql(query.chatId, {
        identifier: "chat_identifier",
        guid: "guid"
      });
      conditions.push(match.sql);
      params.push(...match.params);
    }
    if (query.kind === "group") {
      conditions.push("style = ?");
      params.push(CHAT_STYLE_GROUP);
    } else if (query.kind === "dm") {
      conditions.push("style = ?");
      params.push(CHAT_STYLE_DM);
    }
    if (query.service) {
      conditions.push("service_name = ?");
      params.push(query.service);
    }
    if (query.isArchived === true) {
      conditions.push("is_archived = 1");
    } else if (query.isArchived === false) {
      conditions.push("is_archived = 0");
    }
    if (query.hasUnread) {
      conditions.push("unread_count > 0");
    }
    if (query.search) {
      const escaped = escapeLikePattern(query.search);
      conditions.push("(display_name LIKE ? ESCAPE '\\' OR chat_identifier LIKE ? ESCAPE '\\')");
      params.push(`%${escaped}%`, `%${escaped}%`);
    }
    const where = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
    let orderBy = "";
    if (query.sortBy === "recent") {
      orderBy = "ORDER BY (last_date IS NULL), last_date DESC";
    } else if (query.sortBy === "name") {
      orderBy = "ORDER BY (display_name IS NULL), display_name ASC";
    }
    let limitClause = "";
    if (query.limit != null && query.limit > 0) {
      limitClause = "LIMIT ?";
      params.push(query.limit);
    }
    return {
      sql: `
                WITH chat_stats AS (
                    SELECT
                        chat_message_join.chat_id,
                        MAX(message.date) AS last_date,
                        SUM(
                            CASE
                                WHEN message.is_read = 0 AND message.is_from_me = 0 THEN 1
                                ELSE 0
                            END
                        ) AS unread_count
                    FROM chat_message_join
                    INNER JOIN message ON message.ROWID = chat_message_join.message_id
                    GROUP BY chat_message_join.chat_id
                ),
                enriched AS (
                    SELECT
                        ${CHAT_FIELDS.join(",\n                        ")}
                    FROM chat
                    LEFT JOIN chat_stats ON chat_stats.chat_id = chat.ROWID
                )
                SELECT *
                FROM enriched
                ${where}
                ${orderBy}
                ${limitClause}
            `,
      params
    };
  },
  buildAttachmentQuery(messageIds) {
    const placeholders = messageIds.map(() => "?").join(",");
    return {
      sql: `
                SELECT
                    ${ATTACHMENT_FIELDS.join(",\n                    ")}
                FROM attachment
                INNER JOIN message_attachment_join ON attachment.ROWID = message_attachment_join.attachment_id
                WHERE message_attachment_join.message_id IN (
                    ${placeholders}
                )
                AND (attachment.hide_attachment IS NULL OR attachment.hide_attachment = 0)
                ORDER BY message_attachment_join.message_id ASC, message_attachment_join.attachment_id ASC
            `,
      params: messageIds
    };
  }
};

// src/domain/attachment.ts
function resolveTransferStatus(code) {
  switch (code) {
    case -1:
    case 0:
      return "pending";
    case 1:
    case 2:
    case 3:
    case 4:
      return "transferring";
    case 5:
      return "complete";
    case 6:
    case 7:
      return "failed";
    default:
      return "unknown";
  }
}

// src/domain/message.ts
function resolveMessageKind(itemType, groupActionType) {
  switch (itemType) {
    case 0:
      return "text";
    case 1:
      return groupActionType === 1 ? "memberRemoved" : "memberAdded";
    case 2:
      return "nameChanged";
    case 3:
      return "groupAction";
    default:
      return "unknown";
  }
}
function resolveExpireStatus(code) {
  switch (code) {
    case 1:
      return "willExpire";
    case 2:
      return "expired";
    default:
      return "active";
  }
}
function resolveShareActivity(code) {
  switch (code) {
    case 0:
      return "none";
    case 1:
      return "pending";
    case 2:
      return "active";
    default:
      return "unknown";
  }
}
function resolveShareDirection(code) {
  switch (code) {
    case 0:
      return "none";
    case 1:
      return "incoming";
    case 2:
      return "outgoing";
    default:
      return "unknown";
  }
}
function resolveScheduleKind(code) {
  switch (code) {
    case 0:
      return "none";
    case 1:
      return "sendLater";
    default:
      return "unknown";
  }
}
function resolveScheduleStatus(code) {
  switch (code) {
    case 0:
      return "none";
    case 1:
      return "pending";
    case 2:
      return "sent";
    case 3:
      return "failed";
    default:
      return "unknown";
  }
}

// src/domain/reaction.ts
var REACTION_KIND_MAP = {
  2e3: "love",
  2001: "like",
  2002: "dislike",
  2003: "laugh",
  2004: "emphasize",
  2005: "question",
  2006: "emoji",
  2007: "sticker"
};
var STANDALONE_REACTION_MAP = {
  1e3: "sticker",
  4e3: "pollVote"
};
function resolveReactionMeta(type) {
  if (type == null || type === 0) {
    return { kind: null, isRemoved: false };
  }
  const standalone = STANDALONE_REACTION_MAP[type];
  if (standalone != null) {
    return { kind: standalone, isRemoved: false };
  }
  const isAdd = type >= 2e3 && type <= 2007;
  const isRemove = type >= 3e3 && type <= 3007;
  if (!isAdd && !isRemove) {
    return { kind: null, isRemoved: false };
  }
  const baseType = isRemove ? type - 1e3 : type;
  const kind = REACTION_KIND_MAP[baseType] ?? null;
  return {
    kind,
    isRemoved: kind != null && isRemove
  };
}

// src/domain/service.ts
var SERVICE_MAP = {
  imessage: "iMessage",
  sms: "SMS",
  rcs: "RCS"
};
function resolveService(raw) {
  if (raw == null) return "unknown";
  return SERVICE_MAP[raw.toLowerCase()] ?? "unknown";
}
function firstAttributedString(decoded) {
  const items = Array.isArray(decoded) ? decoded : [decoded];
  for (const item of items) {
    if (item instanceof typedstream.NSAttributedString && item.string) {
      return item.string;
    }
    if (item !== null && typeof item === "object" && "values" in item && Array.isArray(item.values)) {
      for (const val of item.values) {
        if (val instanceof typedstream.NSAttributedString && val.string) {
          return val.string;
        }
      }
    }
  }
  return null;
}
function extractTextFromAttributedBody(blob) {
  try {
    const buffer = Buffer.isBuffer(blob) ? blob : Buffer.from(blob);
    if (buffer.length === 0) return null;
    const decoded = typedstream.Unarchiver.open(buffer, typedstream.Unarchiver.BinaryDecoding.decodable).decodeAll();
    if (!decoded) return null;
    return firstAttributedString(decoded);
  } catch {
    return null;
  }
}

// src/infra/db/mapper.ts
var HOME_DIR = os.homedir();
function parseNumber(value) {
  if (typeof value === "number" && Number.isFinite(value)) {
    return value;
  }
  if (typeof value === "bigint") {
    const parsed = Number(value);
    return Number.isSafeInteger(parsed) ? parsed : null;
  }
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}
function optionalNumber(value, fieldName) {
  if (value == null) return null;
  const parsed = parseNumber(value);
  if (parsed != null) return parsed;
  throw new Error(`Invalid numeric field: ${fieldName}`);
}
function requireNumber(value, fieldName) {
  const parsed = optionalNumber(value, fieldName);
  if (parsed != null) return parsed;
  throw new Error(`Missing numeric field: ${fieldName}`);
}
function optionalString(value, fieldName) {
  if (value == null) return null;
  if (typeof value === "string") return value;
  throw new Error(`Invalid string field: ${fieldName}`);
}
function optionalNonEmptyString(value, fieldName) {
  const parsed = optionalString(value, fieldName);
  return parsed == null || parsed === "" ? null : parsed;
}
function requireNonEmptyString(value, fieldName) {
  const parsed = optionalNonEmptyString(value, fieldName);
  if (parsed != null) return parsed;
  throw new Error(`Missing string field: ${fieldName}`);
}
function flag(value, fieldName) {
  if (typeof value === "boolean") return value;
  if (value == null) return false;
  const parsed = parseNumber(value);
  if (parsed != null) return parsed !== 0;
  throw new Error(`Invalid boolean field: ${fieldName}`);
}
function optionalDate(value, fieldName) {
  const parsed = optionalNumber(value, fieldName);
  if (parsed == null || parsed === 0) return null;
  return fromMacTimestampNs(parsed);
}
function requireDate(value, fieldName) {
  return fromMacTimestampNs(requireNumber(value, fieldName));
}
function optionalParticipant(value, fieldName) {
  const str = optionalString(value, fieldName);
  if (str == null) return null;
  const trimmed = str.trim();
  return trimmed === "" ? null : trimmed;
}
function resolveMessageChatId(row) {
  return resolvePrimaryChatId(row.chat_guid, row.chat_id, row.chat_service) ?? resolveCloudKitChatId(row.ck_chat_id) ?? resolveParticipantChatId(row.participant, row.service, "message.participant") ?? resolveParticipantChatId(row.destination_caller_id, row.service, "message.destination_caller_id");
}
function resolvePrimaryChatId(guid, identifier, service) {
  const guidValue = optionalNonEmptyString(guid, "message.chat_guid");
  const identifierValue = optionalNonEmptyString(identifier, "message.chat_id");
  const serviceValue = optionalNonEmptyString(service, "message.chat_service");
  if (guidValue == null && identifierValue == null && serviceValue == null) {
    return null;
  }
  const raw = guidValue ?? identifierValue ?? "";
  return raw === "" ? null : ChatId.fromUserInput(raw);
}
function resolveCloudKitChatId(value) {
  const raw = optionalParticipant(value, "message.ck_chat_id");
  return raw == null ? null : ChatId.fromUserInput(raw);
}
function resolveParticipantChatId(participant, service, fieldName) {
  const recipient = optionalParticipant(participant, fieldName);
  if (recipient == null) return null;
  const servicePrefix = parseChatServicePrefix(optionalNonEmptyString(service, "message.service"));
  return servicePrefix ? ChatId.fromDMRecipient(recipient, servicePrefix) : ChatId.fromUserInput(recipient);
}
function resolveMessageText(row) {
  const text = optionalString(row.text, "message.text");
  if (text != null && text !== "") {
    return text;
  }
  if (row.attributedBody == null) {
    return null;
  }
  if (!Buffer.isBuffer(row.attributedBody) && !(row.attributedBody instanceof Uint8Array)) {
    return null;
  }
  const value = extractTextFromAttributedBody(row.attributedBody);
  return value === "" ? null : value;
}
function mapReaction(row) {
  const meta = resolveReactionMeta(optionalNumber(row.associated_message_type, "message.associated_message_type"));
  if (meta.kind == null) {
    return null;
  }
  return {
    kind: meta.kind,
    targetMessageId: optionalNonEmptyString(row.associated_message_guid, "message.associated_message_guid"),
    emoji: optionalNonEmptyString(row.associated_message_emoji, "message.associated_message_emoji"),
    textRange: {
      location: optionalNumber(row.associated_message_range_location, "message.associated_message_range_location") ?? 0,
      length: optionalNumber(row.associated_message_range_length, "message.associated_message_range_length") ?? 0
    },
    isRemoved: meta.isRemoved
  };
}
function mapChatId(row) {
  const guid = requireNonEmptyString(row.guid, "chat.guid");
  return ChatId.fromUserInput(guid).toString();
}
function rowToChat(row) {
  return {
    chatId: mapChatId(row),
    name: optionalNonEmptyString(row.display_name, "chat.display_name"),
    service: resolveService(optionalNonEmptyString(row.service_name, "chat.service_name")),
    kind: resolveChatKind(optionalNumber(row.style, "chat.style")),
    account: optionalNonEmptyString(row.account_login, "chat.account_login"),
    isArchived: flag(row.is_archived, "chat.is_archived"),
    isFiltered: flag(row.is_filtered, "chat.is_filtered"),
    dropsIncomingMessages: flag(row.is_blackholed, "chat.is_blackholed"),
    autoDeletesIncomingMessages: flag(row.is_deleting_incoming_messages, "chat.is_deleting_incoming_messages"),
    lastReadAt: optionalDate(row.last_read_message_timestamp, "chat.last_read_message_timestamp"),
    unreadCount: requireNumber(row.unread_count, "chat.unread_count"),
    lastMessageAt: optionalDate(row.last_date, "chat.last_date")
  };
}
function rowToMessage(row, attachments) {
  const resolved = resolveMessageChatId(row);
  const chatId = resolved?.toString() ?? "";
  const chatKind = resolved == null ? "unknown" : resolved.isGroup ? "group" : "dm";
  const errorCode = optionalNumber(row.error, "message.error") ?? 0;
  const reaction = mapReaction(row);
  return {
    rowId: requireNumber(row.id, "message.id"),
    id: requireNonEmptyString(row.guid, "message.guid"),
    chatId,
    chatKind,
    participant: optionalParticipant(row.participant, "message.participant"),
    service: resolveService(optionalNonEmptyString(row.service, "message.service")),
    text: resolveMessageText(row),
    kind: reaction != null ? "reaction" : resolveMessageKind(
      optionalNumber(row.item_type, "message.item_type"),
      optionalNumber(row.group_action_type, "message.group_action_type")
    ),
    isFromMe: flag(row.is_from_me, "message.is_from_me"),
    isRead: flag(row.is_read, "message.is_read"),
    isSent: flag(row.is_sent, "message.is_sent"),
    isDelivered: flag(row.is_delivered, "message.is_delivered"),
    isDowngraded: flag(row.was_downgraded, "message.was_downgraded"),
    didNotifyRecipient: flag(row.did_notify_recipient, "message.did_notify_recipient"),
    isAutoReply: flag(row.is_auto_reply, "message.is_auto_reply"),
    isSystem: flag(row.is_system_message, "message.is_system_message"),
    isForwarded: flag(row.is_forward, "message.is_forward"),
    isAudioMessage: flag(row.is_audio_message, "message.is_audio_message"),
    isPlayed: flag(row.is_played, "message.is_played"),
    isExpirable: flag(row.is_expirable, "message.is_expirable"),
    hasError: errorCode !== 0,
    errorCode,
    isSpam: flag(row.is_spam, "message.is_spam"),
    isContactKeyVerified: flag(row.is_kt_verified, "message.is_kt_verified"),
    hasUnseenMention: flag(row.has_unseen_mention, "message.has_unseen_mention"),
    wasDeliveredQuietly: flag(row.was_delivered_quietly, "message.was_delivered_quietly"),
    isEmergencySos: flag(row.is_sos, "message.is_sos"),
    isCriticalAlert: flag(row.is_critical, "message.is_critical"),
    isOffGridMessage: flag(row.sent_or_received_off_grid, "message.sent_or_received_off_grid"),
    createdAt: requireDate(row.date, "message.date"),
    deliveredAt: optionalDate(row.date_delivered, "message.date_delivered"),
    readAt: optionalDate(row.date_read, "message.date_read"),
    playedAt: optionalDate(row.date_played, "message.date_played"),
    editedAt: optionalDate(row.date_edited, "message.date_edited"),
    retractedAt: optionalDate(row.date_retracted, "message.date_retracted"),
    recoveredAt: optionalDate(row.date_recovered, "message.date_recovered"),
    replyToMessageId: optionalNonEmptyString(row.reply_to_guid, "message.reply_to_guid"),
    threadRootMessageId: optionalNonEmptyString(row.thread_originator_guid, "message.thread_originator_guid"),
    affectedParticipant: optionalParticipant(row.affected_participant, "message.affected_participant"),
    newGroupName: optionalNonEmptyString(row.group_title, "message.group_title"),
    sendEffect: optionalNonEmptyString(row.expressive_send_style_id, "message.expressive_send_style_id"),
    appBundleId: optionalNonEmptyString(row.balloon_bundle_id, "message.balloon_bundle_id"),
    isInvisibleInkRevealed: flag(row.was_detonated, "message.was_detonated"),
    expireStatus: resolveExpireStatus(optionalNumber(row.expire_state, "message.expire_state")),
    shareActivity: resolveShareActivity(optionalNumber(row.share_status, "message.share_status")),
    shareDirection: resolveShareDirection(optionalNumber(row.share_direction, "message.share_direction")),
    scheduleKind: resolveScheduleKind(optionalNumber(row.schedule_type, "message.schedule_type")),
    scheduleStatus: resolveScheduleStatus(optionalNumber(row.schedule_state, "message.schedule_state")),
    segmentCount: optionalNumber(row.part_count, "message.part_count") ?? 0,
    reaction,
    attachments
  };
}
function resolveAttachmentLocalPath(rawFilename) {
  const filename = optionalNonEmptyString(rawFilename, "attachment.filename");
  if (filename == null) return null;
  if (filename.startsWith("~")) {
    return filename.replace(/^~/, HOME_DIR);
  }
  if (!filename.startsWith("/")) {
    return path.join(HOME_DIR, "Library/Messages/Attachments", filename);
  }
  return filename;
}
function resolveAttachmentFileName(rawTransferName, localPath) {
  const transferName = optionalNonEmptyString(rawTransferName, "attachment.transfer_name");
  if (transferName != null) return transferName;
  if (!localPath) return null;
  const name = path.basename(localPath);
  return name === "" ? null : name;
}
function rowToAttachment(row) {
  const localPath = resolveAttachmentLocalPath(row.filename);
  return {
    id: requireNonEmptyString(row.guid, "attachment.guid"),
    fileName: resolveAttachmentFileName(row.transfer_name, localPath),
    localPath,
    mimeType: optionalNonEmptyString(row.mime_type, "attachment.mime_type") ?? "application/octet-stream",
    uti: optionalNonEmptyString(row.uti, "attachment.uti"),
    sizeBytes: optionalNumber(row.total_bytes, "attachment.total_bytes") ?? 0,
    transferStatus: resolveTransferStatus(optionalNumber(row.transfer_state, "attachment.transfer_state")),
    isOutgoing: flag(row.is_outgoing, "attachment.is_outgoing"),
    isSticker: flag(row.is_sticker, "attachment.is_sticker"),
    isSensitiveContent: flag(row.is_commsafety_sensitive, "attachment.is_commsafety_sensitive"),
    altText: optionalNonEmptyString(row.emoji_image_short_description, "attachment.emoji_image_short_description"),
    createdAt: requireDate(row.created_date, "attachment.created_date")
  };
}
var require2 = module$1.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href)));
var cachedCtor;
function loadModule(specifier, extract) {
  try {
    return extract(require2(specifier));
  } catch (error) {
    const cause = toError(error);
    throw DatabaseError(`Failed to load ${specifier}: ${cause.message}`, cause);
  }
}
function resolveCtor() {
  if (cachedCtor) return cachedCtor;
  cachedCtor = typeof Bun !== "undefined" ? loadModule("bun:sqlite", (m) => m.Database) : loadModule("better-sqlite3", (m) => {
    const mod = m;
    return typeof mod === "function" ? mod : mod.default;
  });
  return cachedCtor;
}
var SqliteClient = class {
  constructor(path, readOnly = true) {
    this.readOnly = readOnly;
    const Ctor = resolveCtor();
    try {
      this.db = new Ctor(path, { readonly: readOnly });
    } catch (error) {
      const cause = toError(error);
      throw DatabaseError(`Failed to open database: ${cause.message}`, cause);
    }
  }
  db;
  closed = false;
  all(sql, params = []) {
    if (this.closed) throw DatabaseError("Database is closed");
    try {
      return this.db.prepare(sql).all(...params);
    } catch (error) {
      const cause = toError(error);
      throw DatabaseError(`Query failed: ${cause.message}`, cause);
    }
  }
  close() {
    if (this.closed) return;
    try {
      this.db.close();
    } catch (error) {
      const cause = toError(error);
      throw DatabaseError(`Failed to close database: ${cause.message}`, cause);
    }
    this.closed = true;
  }
};

// src/infra/db/reader.ts
var ATTACHMENT_QUERY_CHUNK = 500;
var SEARCH_PAGE_SIZE = 200;
var QUERIES_BY_SCHEMA = {
  macos26: macos26Queries
};
function resolveSchemaId(columns) {
  return "macos26";
}
function queryMaxRowId(exec) {
  try {
    const rows = exec("SELECT MAX(ROWID) as max_id FROM message");
    return parseNumber(rows[0]?.max_id) ?? 0;
  } catch (error) {
    throw DatabaseError(`Failed to read max ROWID: ${toErrorMessage(error)}`);
  }
}
function executeMessageQuery(exec, queries, sqlQuery, options = {}) {
  const includeAttachments = options.includeAttachments ?? true;
  let messages;
  try {
    const rows = exec(sqlQuery.sql, sqlQuery.params);
    messages = rows.map((row) => rowToMessage(row, []));
  } catch (error) {
    throw DatabaseError(`Failed to query messages: ${toErrorMessage(error)}`);
  }
  if (!includeAttachments) {
    return messages;
  }
  return mergeAttachments(exec, queries, messages);
}
function queryMessages(exec, queries, input) {
  if (input.search) {
    return searchMessages(exec, queries, input);
  }
  return executeMessageQuery(exec, queries, queries.buildMessageQuery(input));
}
function searchMessages(exec, queries, filter) {
  const search = filter.search?.toLowerCase();
  if (!search) {
    return executeMessageQuery(exec, queries, queries.buildMessageQuery(filter));
  }
  const requestedOffset = filter.offset ?? 0;
  const requestedLimit = filter.limit ?? Number.POSITIVE_INFINITY;
  const requiredMatches = requestedOffset + requestedLimit;
  const pageSize = Math.max(SEARCH_PAGE_SIZE, filter.limit ?? 0);
  const matches = [];
  let scanOffset = 0;
  while (matches.length < requiredMatches) {
    const page = executeMessageQuery(
      exec,
      queries,
      queries.buildMessageQuery({
        ...filter,
        limit: pageSize,
        offset: scanOffset
      }),
      { includeAttachments: false }
    );
    if (page.length === 0) break;
    for (const message of page) {
      if (message.text?.toLowerCase().includes(search) === true) {
        matches.push(message);
      }
    }
    if (page.length < pageSize) break;
    scanOffset += pageSize;
  }
  const end = Number.isFinite(requestedLimit) ? requestedOffset + requestedLimit : void 0;
  return mergeAttachments(exec, queries, matches.slice(requestedOffset, end));
}
function queryChats(exec, queries, input) {
  const sqlQuery = queries.buildChatQuery({
    ...input,
    kind: input.kind ?? "all",
    sortBy: input.sortBy ?? "recent"
  });
  try {
    const rows = exec(sqlQuery.sql, sqlQuery.params);
    return rows.map((row) => rowToChat(row));
  } catch (error) {
    throw DatabaseError(`Failed to list chats: ${toErrorMessage(error)}`);
  }
}
function mergeAttachments(exec, queries, messages) {
  if (messages.length === 0) return messages;
  const attachmentMap = batchGetAttachments(
    exec,
    queries,
    messages.map((message) => message.rowId)
  );
  return messages.map((message) => ({
    ...message,
    attachments: attachmentMap.get(message.rowId) ?? []
  }));
}
function batchGetAttachments(exec, queries, messageIds) {
  const result = /* @__PURE__ */ new Map();
  if (messageIds.length === 0) return result;
  for (let i = 0; i < messageIds.length; i += ATTACHMENT_QUERY_CHUNK) {
    const chunk = messageIds.slice(i, i + ATTACHMENT_QUERY_CHUNK);
    const query = queries.buildAttachmentQuery(chunk);
    try {
      const rows = exec(query.sql, query.params);
      for (const row of rows) {
        const messageId = requireNumber(row.msg_id, "attachment.msg_id");
        const attachment = rowToAttachment(row);
        const existing = result.get(messageId);
        if (existing) {
          existing.push(attachment);
        } else {
          result.set(messageId, [attachment]);
        }
      }
    } catch (error) {
      throw DatabaseError(`Failed to query attachments: ${toErrorMessage(error)}`);
    }
  }
  return result;
}
var MessagesDatabaseReader = class extends SqliteClient {
  queries;
  exec;
  constructor(path) {
    super(path, true);
    this.exec = (sql, params) => this.all(sql, params ?? []);
    this.queries = QUERIES_BY_SCHEMA[this.detectSchemaId()];
  }
  /** Get the current maximum message ROWID. */
  async getMaxRowId() {
    return queryMaxRowId(this.exec);
  }
  /** Query messages with optional filters. */
  async getMessages(query = {}) {
    return queryMessages(this.exec, this.queries, query);
  }
  /** Query messages newer than the given ROWID, ordered ascending. */
  async getMessagesSinceRowId(sinceRowId, query = {}) {
    return queryMessages(this.exec, this.queries, {
      ...query,
      sinceRowId,
      orderByRowIdAsc: true
    });
  }
  /** List chats with optional filters and sorting. */
  async listChats(query = {}) {
    return queryChats(this.exec, this.queries, query);
  }
  /** Search messages by text content (application-level filtering). */
  async searchMessages(query) {
    return queryMessages(this.exec, this.queries, query);
  }
  detectSchemaId() {
    const rows = this.all("PRAGMA table_info('message')");
    rows.map((row) => row.name).filter((name) => typeof name === "string" && name !== "");
    return resolveSchemaId();
  }
};
var BATCH_LIMIT = 100;
var MessageWatchSource = class {
  database;
  databasePath;
  walPath;
  walFilename;
  watchFactory;
  onBatch;
  onCheckpoint;
  onError;
  debug;
  isRunning = false;
  lastRowId = -1;
  triggerResolve = null;
  hasPendingTrigger = false;
  walWatcher = null;
  dirWatcher = null;
  constructor(options) {
    this.database = options.database;
    this.databasePath = options.databasePath;
    this.walPath = `${this.databasePath}-wal`;
    this.walFilename = `${path.basename(this.databasePath)}-wal`;
    this.watchFactory = options.watchFactory ?? ((path, listener) => fs.watch(path, listener));
    this.onBatch = options.onBatch;
    this.onCheckpoint = options.onCheckpoint;
    this.onError = options.onError;
    this.debug = options.debug ?? false;
  }
  // -----------------------------------------------
  // Lifecycle
  // -----------------------------------------------
  /** Start watching for new messages. */
  async start() {
    if (this.isRunning) return;
    this.lastRowId = await this.database.getMaxRowId();
    this.isRunning = true;
    try {
      this.ensureWatching();
      this.trigger();
      void this.consumeLoop();
    } catch (error) {
      this.isRunning = false;
      this.detachWALWatcher();
      this.detachDirWatcher();
      throw error;
    }
  }
  /** Stop watching and release all resources. */
  stop() {
    if (!this.isRunning) return;
    this.isRunning = false;
    this.detachWALWatcher();
    this.detachDirWatcher();
    if (this.triggerResolve) {
      this.triggerResolve();
      this.triggerResolve = null;
    }
    this.hasPendingTrigger = false;
  }
  // -----------------------------------------------
  // Single-slot semaphore
  // -----------------------------------------------
  trigger() {
    if (this.triggerResolve) {
      this.triggerResolve();
      this.triggerResolve = null;
      return;
    }
    this.hasPendingTrigger = true;
  }
  waitForTrigger() {
    if (this.hasPendingTrigger) {
      this.hasPendingTrigger = false;
      return Promise.resolve();
    }
    return new Promise((resolve2) => {
      this.triggerResolve = resolve2;
    });
  }
  // -----------------------------------------------
  // Consumer loop
  // -----------------------------------------------
  async consumeLoop() {
    while (this.isRunning) {
      await this.waitForTrigger();
      if (!this.isRunning) break;
      try {
        while (await this.processBatch()) {
          if (!this.isRunning) break;
        }
      } catch (error) {
        this.handleError(error);
      }
    }
  }
  async processBatch() {
    const messages = await this.database.getMessagesSinceRowId(this.lastRowId, { limit: BATCH_LIMIT });
    if (messages.length > 0) {
      if (!this.isRunning) return false;
      await this.onBatch?.(messages);
      const last = messages[messages.length - 1];
      if (last) this.lastRowId = last.rowId;
    }
    try {
      this.onCheckpoint?.(/* @__PURE__ */ new Date());
    } catch (error) {
      this.handleError(error);
    }
    return messages.length === BATCH_LIMIT;
  }
  // -----------------------------------------------
  // WAL / directory watcher state machine
  // -----------------------------------------------
  ensureWatching() {
    this.attachWALWatcher();
    if (!this.walWatcher) this.attachDirWatcher();
    if (!this.walWatcher && !this.dirWatcher) {
      throw new Error("Failed to start watcher: neither WAL file nor directory watch could be established");
    }
  }
  attachWALWatcher() {
    if (this.walWatcher) return;
    try {
      const watcher = this.watchFactory(this.walPath, (eventType) => {
        if (eventType === "change") {
          this.trigger();
        } else if (eventType === "rename") {
          this.detachWALWatcher();
          this.recoverOrStop("WAL file rotated");
        }
      });
      watcher.on("error", () => {
        this.detachWALWatcher();
        this.recoverOrStop("WAL watcher failed");
      });
      this.walWatcher = watcher;
      this.detachDirWatcher();
    } catch {
    }
  }
  detachWALWatcher() {
    if (!this.walWatcher) return;
    this.walWatcher.close();
    this.walWatcher = null;
  }
  attachDirWatcher() {
    if (this.dirWatcher) return;
    try {
      const dir = path.dirname(this.databasePath);
      const watcher = this.watchFactory(dir, (_eventType, filename) => {
        const name = typeof filename === "string" ? filename : Buffer.isBuffer(filename) ? filename.toString("utf8") : null;
        if (name !== this.walFilename) return;
        this.attachWALWatcher();
        this.trigger();
      });
      watcher.on("error", () => {
        this.detachDirWatcher();
        this.recoverOrStop("Directory watcher failed");
      });
      this.dirWatcher = watcher;
    } catch {
    }
  }
  detachDirWatcher() {
    if (!this.dirWatcher) return;
    this.dirWatcher.close();
    this.dirWatcher = null;
  }
  // -----------------------------------------------
  // Error handling
  // -----------------------------------------------
  recoverOrStop(context) {
    try {
      this.ensureWatching();
    } catch (error) {
      this.stop();
      this.handleError(new Error(`${context}: ${toError(error).message}`));
    }
  }
  handleError(error) {
    const err = toError(error);
    if (this.debug) {
      console.error("[WatchSource] Error:", err);
    }
    try {
      this.onError?.(err);
    } catch (callbackError) {
      if (this.debug) {
        console.error("[WatchSource] Error handler failed:", toError(callbackError));
      }
    }
  }
};

// src/domain/validate.ts
var SEND_LIMITS = {
  maxRecipientLength: 320,
  maxTextLength: 1e5,
  maxAttachmentsPerMessage: 50
};
var MAX_RECIPIENT_LENGTH = SEND_LIMITS.maxRecipientLength;
var PHONE_REGEX = /^\+?[\d\s\-()]+$/;
var EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/;
var URL_REGEX = /^https?:\/\/.+$/;
var MIN_PHONE_DIGITS = 7;
var MAX_PHONE_DIGITS = 15;
function validateRecipient(recipient) {
  if (recipient === "") {
    throw SendError("Recipient cannot be empty");
  }
  if (recipient.length > MAX_RECIPIENT_LENGTH) {
    throw SendError(`Recipient exceeds maximum length of ${MAX_RECIPIENT_LENGTH} characters`);
  }
  if (EMAIL_REGEX.test(recipient)) {
    return recipient;
  }
  if (PHONE_REGEX.test(recipient)) {
    const digitCount = recipient.replace(/\D/g, "").length;
    if (digitCount >= MIN_PHONE_DIGITS && digitCount <= MAX_PHONE_DIGITS) {
      return recipient;
    }
  }
  const preview = recipient.length > 80 ? `${recipient.slice(0, 80)}\u2026` : recipient;
  throw SendError(`Invalid recipient format: "${preview}"`);
}
function isURL(value) {
  return URL_REGEX.test(value);
}
var MAX_TEXT_LENGTH = SEND_LIMITS.maxTextLength;
var MAX_ATTACHMENTS_PER_MESSAGE = SEND_LIMITS.maxAttachmentsPerMessage;
function validateMessageContent(text, attachments) {
  const hasText = text != null && text !== "";
  const hasAttachments = attachments != null && attachments.length > 0;
  if (!hasText && !hasAttachments) {
    throw SendError("Message must have text or at least one attachment");
  }
  if (hasText && text.length > MAX_TEXT_LENGTH) {
    throw SendError(`Text exceeds maximum length of ${MAX_TEXT_LENGTH} characters`);
  }
  if (hasAttachments && attachments.length > MAX_ATTACHMENTS_PER_MESSAGE) {
    throw SendError(`Too many attachments (max ${MAX_ATTACHMENTS_PER_MESSAGE})`);
  }
  return { hasText, hasAttachments };
}

// src/domain/routing.ts
function resolveTarget(input) {
  if (input === "") {
    throw SendError("Target cannot be empty");
  }
  const trimmed = input.trim();
  if (trimmed === "") {
    throw SendError("Target cannot be empty");
  }
  const chatId = ChatId.fromUserInput(trimmed);
  if (chatId.isGroup) {
    return { kind: "group", chatId };
  }
  if (!trimmed.includes(";")) {
    const validated = validateRecipient(trimmed);
    return { kind: "dm", recipient: validated, chatIdHint: trimmed };
  }
  const recipient = chatId.extractRecipient();
  if (recipient != null) {
    const validated = validateRecipient(recipient);
    return { kind: "dm", recipient: validated, chatIdHint: trimmed };
  }
  return { kind: "group", chatId };
}

// src/utils/async.ts
function delay(ms, signal) {
  return new Promise((resolve2, reject) => {
    if (signal?.aborted) {
      reject(signal.reason ?? new Error("Aborted"));
      return;
    }
    const cleanup = () => {
      clearTimeout(id);
      signal?.removeEventListener("abort", onAbort);
    };
    const onAbort = () => {
      cleanup();
      reject(signal?.reason ?? new Error("Aborted"));
    };
    const id = setTimeout(() => {
      cleanup();
      resolve2();
    }, ms);
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}
async function retry(fn, options = {}) {
  const attempts = Math.max(1, Math.trunc(options.attempts ?? 3));
  const baseDelay = Math.max(0, options.delay ?? 1e3);
  const backoff = options.backoff ?? true;
  const maxDelay = Math.max(baseDelay, options.maxDelay ?? 3e4);
  const { signal } = options;
  let lastError;
  for (let attempt = 0; attempt < attempts; attempt++) {
    if (signal?.aborted) throw signal.reason ?? new Error("Aborted");
    try {
      return await fn();
    } catch (err) {
      lastError = err instanceof Error ? err : new Error(String(err));
      if (signal?.aborted) break;
      if (attempt < attempts - 1) {
        const exponential = backoff ? baseDelay * 2 ** attempt : baseDelay;
        const capped = Math.min(exponential, maxDelay);
        await delay(Math.random() * capped, signal);
      }
    }
  }
  throw lastError;
}
var Semaphore = class {
  constructor(limit) {
    this.limit = limit;
    if (limit <= 0) throw new Error("Concurrency limit must be greater than 0");
  }
  running = 0;
  waiting = [];
  /** Acquire a slot. Returns a release function. */
  async acquire(signal) {
    if (signal?.aborted) throw signal.reason ?? new Error("Aborted");
    while (this.running >= this.limit) {
      await new Promise((resolve2, reject) => {
        const onAbort = () => {
          const idx = this.waiting.indexOf(wrappedResolve);
          if (idx !== -1) this.waiting.splice(idx, 1);
          reject(signal?.reason ?? new Error("Aborted"));
        };
        const wrappedResolve = () => {
          signal?.removeEventListener("abort", onAbort);
          resolve2();
        };
        signal?.addEventListener("abort", onAbort, { once: true });
        this.waiting.push(wrappedResolve);
      });
    }
    this.running++;
    let released = false;
    return () => {
      if (released) return;
      released = true;
      this.running--;
      this.waiting.shift()?.();
    };
  }
  /** Execute fn within a concurrency slot. */
  async run(fn, signal) {
    const release2 = await this.acquire(signal);
    try {
      return await fn();
    } finally {
      release2();
    }
  }
};
function requireMacOS() {
  if (process.platform !== "darwin") {
    throw PlatformError("Only macOS is supported");
  }
}
function getDefaultDatabasePath() {
  let home;
  try {
    home = os.homedir();
  } catch (cause) {
    throw PlatformError("Unable to resolve user home directory", toError(cause));
  }
  return path.join(home, "Library", "Messages", "chat.db");
}
function getDarwinMajorVersion(osRelease = os.release()) {
  const value = Number.parseInt(osRelease.split(".")[0] ?? "", 10);
  return Number.isFinite(value) ? value : 0;
}
function detectChatServicePrefix(darwinMajor = getDarwinMajorVersion()) {
  return darwinMajor >= 25 ? "any" : "iMessage";
}
var execFileAsync = util.promisify(child_process.execFile);
var DEFAULT_TIMEOUT_MS = 3e4;
function runViaStdin(script, timeoutMs) {
  return new Promise((resolve2, reject) => {
    const child = child_process.spawn("osascript", ["-"], {
      stdio: ["pipe", "pipe", "pipe"],
      timeout: timeoutMs
    });
    let stdout = "";
    let stderr = "";
    let killed = false;
    child.stdout?.on("data", (data) => {
      stdout += data.toString();
    });
    child.stderr?.on("data", (data) => {
      stderr += data.toString();
    });
    child.on("error", (err) => reject(err));
    child.on("close", (code, signal) => {
      if (signal === "SIGTERM") killed = true;
      if (code === 0) {
        resolve2({ stdout, stderr });
      } else {
        const err = new Error(stderr || `osascript exited with code ${code}`);
        err.killed = killed;
        reject(err);
      }
    });
    child.stdin?.write(script);
    child.stdin?.end();
  });
}
async function execAppleScript(script, options) {
  const debug = options?.debug ?? false;
  const timeoutMs = options?.timeout ?? DEFAULT_TIMEOUT_MS;
  if (debug) {
    console.log("[AppleScript] Executing script:\n", script);
  }
  try {
    const { stdout, stderr } = await runViaStdin(script, timeoutMs);
    if (stderr && debug) {
      console.warn("[AppleScript] Warning:", stderr);
    }
    if (debug) {
      console.log("[AppleScript] Success:", stdout || "(no output)");
    }
    return stdout.trim();
  } catch (error) {
    const err = error instanceof Error ? error : new Error(String(error));
    const msg = err.message;
    if ("killed" in err && err.killed || msg.includes("timeout")) {
      throw new Error(`AppleScript execution timeout (${timeoutMs}ms) - may be slow network or large file`);
    }
    if (msg.includes("Can't get buddy")) {
      throw new Error("Recipient not found or not added to iMessage contacts");
    }
    if (msg.includes("Can't send")) {
      throw new Error(
        "Send failed - please check: 1) Is Messages signed in to iMessage, 2) Is recipient correct, 3) Network connection"
      );
    }
    if (debug) {
      console.error("[AppleScript] Error details:", err);
    }
    throw new Error(`AppleScript execution failed: ${msg}`);
  }
}
async function checkMessagesApp() {
  try {
    await execFileAsync("pgrep", ["-x", "Messages"], { timeout: DEFAULT_TIMEOUT_MS });
    return true;
  } catch {
    return false;
  }
}
function escapeAppleScriptString(str) {
  const escapeMap = {
    "\\": "\\\\",
    '"': '\\"',
    "\n": "\\n",
    "\r": "\\r",
    "	": "\\t"
  };
  return str.replace(/\0/g, "").replace(/[\\\n\r\t"]/g, (char) => escapeMap[char] || char);
}
function buildSendScript(params) {
  const { method, identifier, text, attachmentPath } = params;
  const escapedId = escapeAppleScriptString(identifier);
  const targetSetup = method === "buddy" ? `    set targetService to 1st service whose service type = iMessage
    set targetBuddy to buddy "${escapedId}" of targetService` : `    set targetChat to chat id "${escapedId}"`;
  const targetVar = method === "buddy" ? "targetBuddy" : "targetChat";
  const bodyLines = [];
  if (text) {
    const escapedText = escapeAppleScriptString(text);
    bodyLines.push(`    send "${escapedText}" to ${targetVar}`);
  }
  if (attachmentPath) {
    bodyLines.push(buildAttachmentSnippet(attachmentPath, targetVar));
  }
  const body = bodyLines.join("\n\n");
  return {
    script: `tell application "Messages"
${targetSetup}

${body}
end tell`
  };
}
function needsSandboxBypass(filePath) {
  const home = os.homedir();
  const safeDirs = [path.join(home, "Pictures"), path.join(home, "Downloads"), path.join(home, "Documents")];
  return !safeDirs.some((dir) => filePath.startsWith(`${dir}/`) || filePath === dir);
}
function buildTempFilenameParts(filePath) {
  const ext = path.extname(filePath);
  const safeExt = ext.replace(/[^a-zA-Z0-9.]/g, "");
  const base = path.basename(filePath, ext);
  const safeBase = base.replace(/[^a-zA-Z0-9._-]/g, "_").replace(/X/g, "_").slice(0, 60);
  const basePart = safeBase ? `${safeBase}_` : "";
  return { template: `imsg_temp_${basePart}XXXXXXXXXX`, ext: safeExt };
}
function calculateFileDelay(filePath) {
  try {
    const stats = fs.statSync(filePath);
    const sizeInMB = stats.size / (1024 * 1024);
    if (sizeInMB < 1) return 2;
    if (sizeInMB < 10) return 3;
    return 5;
  } catch {
    return 3;
  }
}
function buildAttachmentSnippet(filePath, targetVar) {
  return needsSandboxBypass(filePath) ? buildSandboxBypassSnippet(filePath, targetVar) : buildDirectSendSnippet(filePath, targetVar);
}
function buildSandboxBypassSnippet(filePath, targetVar) {
  const escapedFilePath = escapeAppleScriptString(filePath);
  const { template, ext } = buildTempFilenameParts(filePath);
  const escapedTemplate = escapeAppleScriptString(template);
  const escapedExt = escapeAppleScriptString(ext);
  const fileDelay = calculateFileDelay(filePath);
  return `    set picturesFolder to POSIX path of (path to pictures folder)
    set tmpBase to do shell script "mktemp " & quoted form of (picturesFolder & "${escapedTemplate}")
    set targetPath to tmpBase & "${escapedExt}"
    do shell script "mv " & quoted form of tmpBase & " " & quoted form of targetPath
    do shell script "cat " & quoted form of "${escapedFilePath}" & " > " & quoted form of targetPath & " && chmod 600 " & quoted form of targetPath & " || { rm -f " & quoted form of targetPath & "; exit 1; }"

    set theFile to (POSIX file targetPath) as alias
    send theFile to ${targetVar}
    delay ${fileDelay}`;
}
function buildDirectSendSnippet(filePath, targetVar) {
  const escapedPath = escapeAppleScriptString(filePath);
  const fileDelay = calculateFileDelay(filePath);
  return `    send POSIX file "${escapedPath}" to ${targetVar}
    delay ${fileDelay}`;
}
var execFileAsync2 = util.promisify(child_process.execFile);
var TEMP_DIR = path.join(os.homedir(), "Pictures");
var MAX_DOWNLOAD_SIZE = 50 * 1024 * 1024;
function generateSecureFilename(ext) {
  return `imsg_temp_${crypto$1.randomBytes(8).toString("hex")}${ext}`;
}
function redactUrlForLogs(url) {
  try {
    const parsed = new URL(url);
    const pathname = parsed.pathname.length > 80 ? `${parsed.pathname.slice(0, 80)}...` : parsed.pathname;
    return `${parsed.origin}${pathname}`;
  } catch {
    return url.length > 120 ? `${url.slice(0, 120)}...` : url;
  }
}
function isSupportedImageContentType(contentType) {
  return contentType.startsWith("image/") || contentType.startsWith("application/octet-stream") || contentType === "";
}
async function convertImageToJPEG(inputPath, outputPath) {
  const output = outputPath || path.join(TEMP_DIR, generateSecureFilename(".jpg"));
  try {
    await execFileAsync2("sips", ["-s", "format", "jpeg", inputPath, "--out", output], {
      timeout: 1e4
    });
    if (!fs.existsSync(output)) {
      throw new Error("Converted file does not exist");
    }
    return output;
  } catch (error) {
    throw SendError(`Image format conversion failed: ${toErrorMessage(error)}`);
  }
}
async function downloadImage(url, options = {}) {
  const {
    timeout = 15e3,
    maxRetries = 2,
    retryDelay = 1e3,
    maxSize = MAX_DOWNLOAD_SIZE,
    debug = false,
    signal
  } = options;
  const safeUrl = redactUrlForLogs(url);
  if (debug) {
    console.log(`[Download] ${safeUrl}`);
  }
  try {
    return await retry(() => fetchAndSave(url, timeout, maxSize, debug, signal), {
      attempts: maxRetries + 1,
      delay: retryDelay,
      backoff: true,
      signal
    });
  } catch (error) {
    throw SendError(`Download failed (${maxRetries + 1} attempts): ${toErrorMessage(error)}
URL: ${safeUrl}`);
  }
}
async function convertToCompatibleFormat(filePath) {
  const ext = path.extname(filePath).slice(1).toLowerCase();
  if (!ext || !["avif", "webp"].includes(ext)) {
    return { path: filePath, converted: false };
  }
  const tail = path.basename(filePath);
  if (!tail) {
    return { path: filePath, converted: false };
  }
  const fileName = tail.replace(/\.(avif|webp)$/i, ".jpg");
  const isOurTemp = fileName.startsWith("imsg_temp_");
  const output = isOurTemp ? path.join(TEMP_DIR, fileName) : path.join(TEMP_DIR, generateSecureFilename(".jpg"));
  const converted = await convertImageToJPEG(filePath, output);
  return { path: converted, converted: true };
}
async function fetchAndSave(url, timeout, maxSize, debug, signal) {
  const fetchSignal = signal ? AbortSignal.any([signal, AbortSignal.timeout(timeout)]) : AbortSignal.timeout(timeout);
  const response = await fetch(url, {
    signal: fetchSignal,
    headers: {
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36",
      Accept: "image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8",
      "Accept-Encoding": "gzip, deflate, br"
    }
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status} ${response.statusText}`);
  }
  const contentLength = response.headers.get("content-length");
  if (contentLength && Number.parseInt(contentLength, 10) > maxSize) {
    throw new Error(`File too large: ${contentLength} bytes (max ${maxSize})`);
  }
  const contentType = response.headers.get("content-type")?.toLowerCase() || "";
  if (!isSupportedImageContentType(contentType)) {
    throw new Error(`Unsupported content type: ${contentType}`);
  }
  const buffer = await readResponseWithLimit(response, maxSize);
  if (contentType.includes("avif") || contentType.includes("webp")) {
    const ext2 = contentType.includes("avif") ? ".avif" : ".webp";
    const tempPath = path.join(TEMP_DIR, generateSecureFilename(ext2));
    fs.writeFileSync(tempPath, buffer, { flag: "wx", mode: 384 });
    let converted;
    try {
      converted = await convertImageToJPEG(tempPath);
    } catch (error) {
      try {
        fs.unlinkSync(tempPath);
      } catch {
      }
      throw error;
    }
    try {
      fs.unlinkSync(tempPath);
    } catch {
    }
    if (debug) console.log(`[Download] Converted ${ext2} -> .jpg`);
    return converted;
  }
  const extMap = { png: ".png", gif: ".gif", svg: ".svg", bmp: ".bmp" };
  const ext = Object.entries(extMap).find(([key]) => contentType.includes(key))?.[1] || ".jpg";
  const path$1 = path.join(TEMP_DIR, generateSecureFilename(ext));
  fs.writeFileSync(path$1, buffer, { flag: "wx", mode: 384 });
  return path$1;
}
async function readResponseWithLimit(response, maxSize) {
  if (!response.body) {
    const ab = await response.arrayBuffer();
    if (ab.byteLength > maxSize) throw new Error(`Download exceeded max size (${maxSize} bytes)`);
    return Buffer.from(ab);
  }
  const reader = response.body.getReader();
  const chunks = [];
  let totalBytes = 0;
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      totalBytes += value.byteLength;
      if (totalBytes > maxSize) {
        reader.cancel();
        throw new Error(`Download exceeded max size (${maxSize} bytes)`);
      }
      chunks.push(value);
    }
  } finally {
    reader.releaseLock();
  }
  return Buffer.concat(chunks);
}

// src/infra/outgoing/tracker.ts
var MessagePromiseRejection = class extends Error {
  constructor(reason) {
    super(reason);
    this.name = this.constructor.name;
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }
};
var MessagePromise = class {
  promise;
  resolvePromise;
  rejectPromise;
  chatId;
  coreChatId;
  text;
  attachmentName;
  isAttachment;
  sentAt;
  debug;
  isResolved = false;
  errored = false;
  error = null;
  timeoutHandle;
  constructor(options) {
    this.debug = options.debug ?? false;
    this.chatId = options.chatId;
    this.coreChatId = ChatId.fromUserInput(options.chatId).coreIdentifier;
    this.text = normalizeText(options.text ?? "");
    this.attachmentName = options.attachmentName ?? null;
    this.isAttachment = options.isAttachment;
    this.sentAt = typeof options.sentAt === "number" ? options.sentAt : options.sentAt.getTime();
    this.promise = new Promise((resolve2, reject) => {
      this.resolvePromise = resolve2;
      this.rejectPromise = reject;
    });
    this.promise.catch((err) => {
      this.errored = true;
      this.error = err instanceof Error ? err : new Error(String(err));
    });
    const timeout = options.timeout ?? (this.isAttachment ? 3e4 : 1e4);
    this.timeoutHandle = setTimeout(() => {
      if (!this.isResolved) {
        this.reject(`Message send timeout after ${timeout}ms`);
      }
    }, timeout);
    this.timeoutHandle.unref?.();
  }
  // -----------------------------------------------
  // Resolve / Reject
  // -----------------------------------------------
  resolve(message) {
    if (this.isResolved) return;
    this.isResolved = true;
    clearTimeout(this.timeoutHandle);
    this.resolvePromise(message);
  }
  reject(reason) {
    if (this.isResolved) return;
    this.isResolved = true;
    clearTimeout(this.timeoutHandle);
    this.rejectPromise(new MessagePromiseRejection(reason));
  }
  // -----------------------------------------------
  // Matching
  // -----------------------------------------------
  /** Check whether a database message corresponds to this outgoing send. */
  matches(message) {
    const timeDiff = message.createdAt.getTime() - this.sentAt;
    if (timeDiff < -5e3 || timeDiff > 12e4) {
      if (this.debug) console.log("[MessagePromise] Time check failed:", { timeDiff });
      return false;
    }
    if (!this.matchesChatId(message.chatId)) {
      if (this.debug) {
        console.log("[MessagePromise] ChatId mismatch:", {
          expected: this.chatId,
          actual: message.chatId
        });
      }
      return false;
    }
    if (this.isAttachment) {
      if (!message.attachments || message.attachments.length === 0) {
        if (this.debug) console.log("[MessagePromise] No attachments in message");
        return false;
      }
      if (this.attachmentName) {
        const expected = filenameWithoutExt(this.attachmentName).toLowerCase();
        const match = message.attachments.some((att) => {
          const actual = filenameWithoutExt(att.fileName).toLowerCase();
          return actual === expected;
        });
        if (this.debug) {
          console.log("[MessagePromise] Attachment match:", {
            expected,
            actual: message.attachments.map((a) => filenameWithoutExt(a.fileName).toLowerCase()),
            match
          });
        }
        return match;
      }
      return true;
    }
    const messageText = normalizeText(message.text ?? "");
    const matches = this.text === messageText;
    if (this.debug) {
      console.log("[MessagePromise] Text match:", { expected: this.text, actual: messageText, matches });
    }
    return matches;
  }
  matchesChatId(chatId) {
    if (chatId === this.chatId) return true;
    return ChatId.fromUserInput(chatId).coreIdentifier === this.coreChatId;
  }
};
var OutgoingMessageManager = class {
  byChat = /* @__PURE__ */ new Map();
  debug;
  constructor(debug = false) {
    this.debug = debug;
  }
  add(promise) {
    const key = promise.coreChatId;
    const list = this.byChat.get(key);
    if (list) {
      list.push(promise);
    } else {
      this.byChat.set(key, [promise]);
    }
    if (this.debug) {
      const type = promise.isAttachment ? "attachment" : "message";
      console.log(`[OutgoingManager] Tracking new ${type}: ${promise.chatId}`);
    }
  }
  tryMatch(message) {
    if (!message.isFromMe) return false;
    const key = ChatId.fromUserInput(message.chatId).coreIdentifier;
    const promises = this.byChat.get(key);
    if (!promises) return false;
    for (const promise of promises) {
      if (promise.isResolved) continue;
      if (promise.matches(message)) {
        if (this.debug) {
          console.log(`[OutgoingManager] Matched message: ${message.rowId}`);
        }
        promise.resolve(message);
        return true;
      }
    }
    return false;
  }
  cleanup() {
    const now = Date.now();
    const resolvedCutoff = now - 6e4;
    const staleCutoff = now - 12e4;
    for (const [key, promises] of this.byChat.entries()) {
      const remaining = promises.filter((p) => {
        if (p.isResolved) return p.sentAt > resolvedCutoff;
        return p.sentAt > staleCutoff;
      });
      remaining.length ? this.byChat.set(key, remaining) : this.byChat.delete(key);
    }
  }
  getPendingCount() {
    let count = 0;
    for (const promises of this.byChat.values()) {
      for (const p of promises) {
        if (!p.isResolved) count++;
      }
    }
    return count;
  }
  rejectAll(reason = "SDK closed") {
    for (const promises of this.byChat.values()) {
      for (const promise of promises) {
        if (!promise.isResolved) {
          promise.reject(reason);
        }
      }
    }
    this.byChat.clear();
  }
};
function normalizeText(text) {
  return text.replace(/\s+/g, " ").trim().toLowerCase();
}
function filenameWithoutExt(filename) {
  const base = (filename ?? "").split(/[\\/]/).pop() ?? "";
  const lastDot = base.lastIndexOf(".");
  return lastDot === -1 ? base : base.substring(0, lastDot);
}

// src/infra/outgoing/sender.ts
var SEND_TIMEOUT_MS = 3e4;
var RETRY_MAX_RETRIES = 2;
var RETRY_DELAY_MS = 1500;
function resolveRequestedService(service) {
  if (service == null || service === "iMessage") {
    return "iMessage";
  }
  throw SendError(`Outbound send service "${service}" is not supported`);
}
var MessageSender = class {
  debug;
  maxRetries;
  retryDelay;
  semaphore;
  sendTimeoutMs;
  outgoingManager;
  chatServicePrefix;
  constructor(options = {}) {
    this.debug = options.debug ?? false;
    this.maxRetries = options.retryAttempts ?? RETRY_MAX_RETRIES;
    this.retryDelay = options.retryDelay ?? RETRY_DELAY_MS;
    this.semaphore = options.semaphore ?? null;
    this.sendTimeoutMs = options.timeout ?? SEND_TIMEOUT_MS;
    this.outgoingManager = options.outgoingManager ?? null;
    this.chatServicePrefix = detectChatServicePrefix();
  }
  // -----------------------------------------------
  // Public API
  // -----------------------------------------------
  /** Send a message to a recipient or group chat. */
  async send(request) {
    return this.executeSend(this.createSendJob(request));
  }
  // -----------------------------------------------
  // Send Pipeline
  // -----------------------------------------------
  async executeSend(job) {
    const { target, chatId, service, signal, timeout, label, text, attachments } = job;
    const effectiveTimeout = timeout ?? this.sendTimeoutMs;
    const task = async () => {
      this.checkAbortSignal(signal);
      let hasText;
      try {
        const validation = validateMessageContent(text, attachments);
        hasText = validation.hasText;
      } catch (error) {
        throw SendError(toErrorMessage(error));
      }
      try {
        this.checkAbortSignal(signal);
        await this.checkMessagesEnvironment();
        const paths = await this.prepareAttachments(attachments, signal);
        const sentAt = /* @__PURE__ */ new Date();
        const messagePromise = this.createTrackingPromise(chatId, text, hasText, paths, sentAt);
        await this.executeAppleScripts(target, text, hasText, paths, effectiveTimeout, signal);
        const confirmedMessage = await this.awaitConfirmation(messagePromise);
        this.outgoingManager?.cleanup();
        return {
          chatId,
          to: target.identifier,
          service,
          sentAt,
          message: confirmedMessage
        };
      } catch (error) {
        const errorMsg = toErrorMessage(error);
        const context = `[${label}] [Text: ${hasText ? "yes" : "no"}] [Attachments: ${attachments.length}]`;
        const cause = error instanceof Error ? error : void 0;
        if (error instanceof IMessageError) {
          throw SendError(`${errorMsg} ${context}`, cause);
        }
        throw SendError(`Send failed ${context}: ${errorMsg}`, cause);
      }
    };
    return this.semaphore ? await this.semaphore.run(task, job.signal) : await task();
  }
  createSendJob(request) {
    const { to, text, attachments = [], signal, timeout } = request;
    const resolved = resolveTarget(String(to));
    const resolvedAttachments = [...attachments];
    const service = resolveRequestedService(request.service);
    if (resolved.kind === "group") {
      const parsed = resolved.chatId;
      try {
        parsed.validate();
      } catch (error) {
        throw SendError(toErrorMessage(error));
      }
      const normalizedId = parsed.buildGroupGuid(this.chatServicePrefix);
      return {
        target: { method: "chat", identifier: normalizedId },
        chatId: to,
        service,
        text,
        attachments: resolvedAttachments,
        signal,
        timeout,
        label: `Group: ${normalizedId}`
      };
    }
    const recipient = resolved.recipient;
    return {
      target: { method: "buddy", identifier: recipient },
      chatId: ChatId.fromDMRecipient(recipient, this.chatServicePrefix).toString(),
      service,
      text,
      attachments: resolvedAttachments,
      signal,
      timeout,
      label: `To: ${recipient}`
    };
  }
  // -----------------------------------------------
  // Pipeline Steps
  // -----------------------------------------------
  createTrackingPromise(chatId, text, hasText, paths, sentAt) {
    if (!this.outgoingManager) return null;
    let messagePromise = null;
    if (paths.length > 0) {
      messagePromise = new MessagePromise({
        chatId,
        text,
        attachmentName: paths[0] ? path.basename(paths[0]) : void 0,
        isAttachment: true,
        sentAt
      });
    } else if (hasText) {
      messagePromise = new MessagePromise({
        chatId,
        text,
        isAttachment: false,
        sentAt
      });
    }
    if (messagePromise) {
      this.outgoingManager.add(messagePromise);
    }
    return messagePromise;
  }
  async executeAppleScripts(target, text, hasText, resolvedPaths, timeoutMs, signal) {
    const { method, identifier: id } = target;
    const descPrefix = method === "chat" ? `group ${id}` : id;
    if (hasText && resolvedPaths.length > 0) {
      const firstAttachment = resolvedPaths[0];
      const { script } = buildSendScript({ method, identifier: id, text, attachmentPath: firstAttachment });
      await this.executeWithRetry(script, `Send text and attachment to ${descPrefix}`, timeoutMs, signal);
      for (let i = 1; i < resolvedPaths.length; i++) {
        const path = resolvedPaths[i];
        if (!path) continue;
        const { script: attachScript } = buildSendScript({
          method,
          identifier: id,
          attachmentPath: path
        });
        await this.executeWithRetry(
          attachScript,
          `Send attachment ${i + 1}/${resolvedPaths.length}`,
          timeoutMs,
          signal
        );
        if (i < resolvedPaths.length - 1) await delay(500, signal);
      }
    } else if (hasText) {
      const { script } = buildSendScript({ method, identifier: id, text });
      await this.executeWithRetry(script, `Send text to ${descPrefix}`, timeoutMs, signal);
    } else {
      for (let i = 0; i < resolvedPaths.length; i++) {
        const path = resolvedPaths[i];
        if (!path) continue;
        const { script } = buildSendScript({ method, identifier: id, attachmentPath: path });
        await this.executeWithRetry(
          script,
          `Send attachment ${i + 1}/${resolvedPaths.length} to ${descPrefix}`,
          timeoutMs,
          signal
        );
        if (i < resolvedPaths.length - 1) await delay(500, signal);
      }
    }
  }
  async awaitConfirmation(messagePromise) {
    if (!messagePromise) return void 0;
    try {
      const message = await messagePromise.promise;
      if (this.debug) {
        console.log("[Sender] Message confirmed in database");
      }
      return message;
    } catch (promiseError) {
      if (this.debug) {
        console.warn("[Sender] Message promise rejected:", promiseError);
      }
      return void 0;
    }
  }
  // -----------------------------------------------
  // Shared Helpers
  // -----------------------------------------------
  async executeWithRetry(script, description, timeoutMs, signal) {
    const totalAttempts = this.maxRetries + 1;
    try {
      await retry(() => execAppleScript(script, { debug: this.debug, timeout: timeoutMs }), {
        attempts: totalAttempts,
        delay: this.retryDelay,
        signal
      });
    } catch (error) {
      throw SendError(`${description} failed after ${totalAttempts} attempts: ${toErrorMessage(error)}`);
    }
  }
  async resolveAttachment(path$1, signal) {
    if (isURL(path$1)) {
      return await downloadImage(path$1, { debug: this.debug, signal });
    }
    const localPath = path.resolve(path$1);
    if (!fs.existsSync(localPath)) {
      const name = path.basename(path$1) || "unknown";
      throw SendError(`File not found: ${name}`);
    }
    const converted = await convertToCompatibleFormat(localPath);
    if (converted.converted && this.debug) {
      const originalFile = path.basename(localPath);
      const convertedFile = path.basename(converted.path);
      console.log(`[Format Conversion] ${originalFile} -> ${convertedFile}`);
    }
    return converted.path;
  }
  checkAbortSignal(signal) {
    if (signal?.aborted) {
      throw SendError("Send cancelled");
    }
  }
  async checkMessagesEnvironment() {
    const isAvailable = await checkMessagesApp();
    if (!isAvailable) {
      throw SendError("Messages app is not running");
    }
  }
  async prepareAttachments(attachments, signal) {
    if (attachments.length === 0) return [];
    if (this.debug) {
      console.log(`[Processing Attachments] Total ${attachments.length} attachments`);
    }
    const resolvedPaths = [];
    for (let i = 0; i < attachments.length; i++) {
      this.checkAbortSignal(signal);
      const attachment = attachments[i];
      if (attachment === void 0) continue;
      if (this.debug) {
        const preview = attachment.length > 80 ? `${attachment.slice(0, 80)}...` : attachment;
        console.log(`[Processing Attachments] ${i + 1}/${attachments.length}: ${preview}`);
      }
      const resolved = await this.resolveAttachment(attachment, signal);
      resolvedPaths.push(resolved);
    }
    return resolvedPaths;
  }
};
var TEMP_FILE_PREFIX = "imsg_temp_";
var TEMP_DIR2 = path.join(os.homedir(), "Pictures");
var DEFAULTS = {
  maxAge: 10 * 60 * 1e3,
  cleanupInterval: 5 * 60 * 1e3
};
var TempFileManager = class {
  config;
  cleanupTimer = null;
  destroyPromise = null;
  isDestroying = false;
  isDestroyed = false;
  constructor(config = {}) {
    this.config = {
      maxAge: config.maxAge ?? DEFAULTS.maxAge,
      cleanupInterval: config.cleanupInterval ?? DEFAULTS.cleanupInterval,
      debug: config.debug ?? false
    };
  }
  // -----------------------------------------------
  // Lifecycle
  // -----------------------------------------------
  /** Start the periodic cleanup timer. */
  start() {
    if (this.isDestroying) {
      throw new Error("TempFileManager is destroying, cannot start");
    }
    if (this.isDestroyed) {
      throw new Error("TempFileManager is destroyed, cannot start");
    }
    if (this.cleanupTimer) return;
    this.cleanup().catch((error) => {
      if (this.config.debug) {
        console.error("[TempFileManager] Startup cleanup failed:", error);
      }
    });
    this.cleanupTimer = setInterval(() => {
      this.cleanup().catch((error) => {
        if (this.config.debug) {
          console.error("[TempFileManager] Periodic cleanup failed:", error);
        }
      });
    }, this.config.cleanupInterval);
    this.cleanupTimer.unref();
    if (this.config.debug) {
      const intervalSec = this.config.cleanupInterval / 1e3;
      const maxAgeSec = this.config.maxAge / 1e3;
      console.log(`[TempFileManager] Started, interval: ${intervalSec}s, maxAge: ${maxAgeSec}s`);
    }
  }
  /** Stop the periodic cleanup timer. */
  stop() {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
  }
  /** Stop the timer and remove expired files. */
  async destroy() {
    if (this.isDestroyed) return;
    if (this.destroyPromise) {
      await this.destroyPromise;
      return;
    }
    this.destroyPromise = (async () => {
      this.isDestroying = true;
      this.stop();
      try {
        await this.removeFiles({ ignoreAge: false });
      } finally {
        this.isDestroyed = true;
        this.isDestroying = false;
        if (this.config.debug) {
          console.log("[TempFileManager] Destroyed");
        }
      }
    })();
    try {
      await this.destroyPromise;
    } finally {
      this.destroyPromise = null;
    }
  }
  // -----------------------------------------------
  // Cleanup
  // -----------------------------------------------
  /** Remove expired temp files. */
  async cleanup() {
    if (this.isDestroyed) return { removed: 0, errors: 0 };
    return this.removeFiles({ ignoreAge: false });
  }
  /** Remove all temp files regardless of age. */
  async cleanupAll() {
    if (this.isDestroyed) return { removed: 0, errors: 0 };
    return this.removeFiles({ ignoreAge: true });
  }
  // -----------------------------------------------
  // Stats
  // -----------------------------------------------
  /** Return current temp file count, running state, and config. */
  getStats() {
    let currentFiles = 0;
    try {
      if (fs.existsSync(TEMP_DIR2)) {
        const files = fs.readdirSync(TEMP_DIR2);
        currentFiles = files.filter((f) => f.startsWith(TEMP_FILE_PREFIX)).length;
      }
    } catch {
    }
    return {
      currentFiles,
      isRunning: this.cleanupTimer !== null,
      config: { ...this.config }
    };
  }
  // -----------------------------------------------
  // Internal
  // -----------------------------------------------
  async removeFiles(options) {
    let removed = 0;
    let errors = 0;
    const now = Date.now();
    try {
      if (!fs.existsSync(TEMP_DIR2)) return { removed, errors };
      const files = fs.readdirSync(TEMP_DIR2);
      for (const file of files) {
        if (!file.startsWith(TEMP_FILE_PREFIX)) continue;
        const filePath = path.join(TEMP_DIR2, file);
        try {
          const stats = fs.lstatSync(filePath);
          if (!stats.isFile()) {
            if (this.config.debug) {
              console.log(`[TempFileManager] Skipping non-regular file: ${file}`);
            }
            continue;
          }
          if (!options.ignoreAge) {
            const fileAge = now - stats.mtimeMs;
            if (fileAge <= this.config.maxAge) continue;
          }
          fs.unlinkSync(filePath);
          removed++;
          if (this.config.debug) {
            console.log(`[TempFileManager] Removed: ${file}`);
          }
        } catch (error) {
          errors++;
          if (this.config.debug) {
            console.error(`[TempFileManager] Failed to remove: ${file}`, error);
          }
        }
      }
      if (this.config.debug && (removed > 0 || errors > 0)) {
        console.log(`[TempFileManager] Done: removed ${removed}, errors ${errors}`);
      }
    } catch (error) {
      if (this.config.debug) {
        console.error("[TempFileManager] Cleanup process error:", error);
      }
    }
    return { removed, errors };
  }
};

// src/infra/plugin/manager.ts
var SEQUENTIAL_HOOKS = /* @__PURE__ */ new Set([
  "onInit",
  "onBeforeMessageQuery",
  "onBeforeChatQuery",
  "onBeforeSend",
  "onError",
  "onDestroy"
]);
var PluginManager = class {
  plugins = [];
  pendingInits = [];
  destroying = false;
  _initialized = false;
  get initialized() {
    return this._initialized;
  }
  // -----------------------------------------------
  // Registration
  // -----------------------------------------------
  /** Register a plugin. Deduplicates by name; late registrations auto-init. */
  use(plugin) {
    if (this.plugins.some((current) => current.name === plugin.name)) {
      throw new Error(`Plugin "${plugin.name}" is already registered`);
    }
    this.plugins.push(plugin);
    if (this._initialized && !this.destroying && plugin.onInit) {
      const initPromise = Promise.resolve().then(() => plugin.onInit?.()).catch((error) => this.reportHookError(plugin.name, "onInit", error));
      this.pendingInits.push(initPromise);
    }
    return this;
  }
  // -----------------------------------------------
  // Lifecycle
  // -----------------------------------------------
  /** Initialize all registered plugins. Idempotent — subsequent calls flush late registrations. */
  async init() {
    if (this._initialized) {
      await this.flushPendingInits();
      return;
    }
    this._initialized = true;
    await this.callHook("onInit");
    await this.flushPendingInits();
  }
  /** Destroy all plugins and reset state. */
  async destroy() {
    if (this.destroying) return;
    this.destroying = true;
    try {
      await this.flushPendingInits();
      await this.callHook("onDestroy");
      this.plugins = [];
      this._initialized = false;
    } finally {
      this.destroying = false;
    }
  }
  // -----------------------------------------------
  // Hook Dispatch
  // -----------------------------------------------
  /** Dispatch a hook to all plugins that implement it. */
  async callHook(hookName, ...args) {
    const pluginsWithHook = this.sortedPluginsWithHook(hookName);
    if (pluginsWithHook.length === 0) return [];
    const errors = SEQUENTIAL_HOOKS.has(hookName) ? await this.dispatchSequential(pluginsWithHook, hookName, args) : await this.dispatchParallel(pluginsWithHook, hookName, args);
    if (errors.length > 0 && hookName !== "onError") {
      for (const { plugin, error } of errors) {
        await this.reportHookError(plugin, hookName, error);
      }
    }
    return errors;
  }
  // -----------------------------------------------
  // Dispatch Strategies
  // -----------------------------------------------
  async dispatchSequential(plugins, hookName, args) {
    const errors = [];
    for (const plugin of plugins) {
      try {
        const hook = plugin[hookName];
        if (hook === void 0) continue;
        const hookFn = hook;
        await Promise.resolve(hookFn(...args));
      } catch (error) {
        errors.push({ plugin: plugin.name, error: toError(error) });
      }
    }
    return errors;
  }
  async dispatchParallel(plugins, hookName, args) {
    const results = await Promise.allSettled(
      plugins.map(async (plugin) => {
        try {
          const hook = plugin[hookName];
          if (hook === void 0) {
            throw new Error(`Invariant: plugin "${plugin.name}" is missing hook ${String(hookName)}`);
          }
          const hookFn = hook;
          await Promise.resolve(hookFn(...args));
          return { plugin: plugin.name, success: true };
        } catch (error) {
          return { plugin: plugin.name, success: false, error: toError(error) };
        }
      })
    );
    const errors = [];
    for (const result of results) {
      if (result.status === "fulfilled" && !result.value.success) {
        errors.push({ plugin: result.value.plugin, error: result.value.error });
      } else if (result.status === "rejected") {
        errors.push({ plugin: "unknown", error: toError(result.reason) });
      }
    }
    return errors;
  }
  // -----------------------------------------------
  // Ordering
  // -----------------------------------------------
  sortedPluginsWithHook(hookName) {
    const withHook = this.plugins.filter((p) => p[hookName]);
    if (withHook.length <= 1) return withHook;
    const pre = [];
    const normal = [];
    const post = [];
    for (const plugin of withHook) {
      if (plugin.order === "pre") pre.push(plugin);
      else if (plugin.order === "post") post.push(plugin);
      else normal.push(plugin);
    }
    return [...pre, ...normal, ...post];
  }
  // -----------------------------------------------
  // Error Reporting
  // -----------------------------------------------
  async flushPendingInits() {
    if (this.pendingInits.length > 0) {
      await Promise.all(this.pendingInits);
      this.pendingInits = [];
    }
  }
  async reportHookError(plugin, hookName, error) {
    const normalizedError = toError(error);
    console.error(`[Plugin ${plugin}] ${hookName} failed:`, normalizedError);
    try {
      await this.callHook("onError", {
        error: normalizedError,
        context: `Plugin ${plugin} - ${String(hookName)}`
      });
    } catch (hookError) {
      console.error(`[Plugin] onError hook failed (suppressed to prevent recursion):`, toError(hookError));
    }
  }
};
var definePlugin = (plugin) => plugin;
function createPluginMessageSink(manager) {
  return {
    async onMessage(message) {
      await manager.callHook("onNewMessage", { message });
    },
    onError(error, context) {
      const ctx = { error, context };
      void manager.callHook("onError", ctx);
    }
  };
}

// src/sdk-bounds.ts
var BOUNDS = {
  maxConcurrentSends: { default: 5, min: 1, max: 50 }
};

// src/sdk.ts
var ERROR_SDK_DESTROYED = "SDK is destroyed";
var ERROR_WATCHER_RUNNING = "Watcher is already running";
var IMessageSDK = class {
  databasePath;
  debug;
  maxConcurrentSends;
  database;
  tempFiles;
  sender;
  plugins;
  outgoing;
  abortController = new AbortController();
  watchSource = null;
  destroyed = false;
  constructor(config = {}) {
    requireMacOS();
    this.databasePath = config.databasePath ?? getDefaultDatabasePath();
    this.debug = config.debug ?? false;
    this.maxConcurrentSends = validateBound(
      config.maxConcurrentSends,
      BOUNDS.maxConcurrentSends,
      "maxConcurrentSends"
    );
    this.database = new MessagesDatabaseReader(this.databasePath);
    this.outgoing = new OutgoingMessageManager(this.debug);
    this.plugins = new PluginManager();
    const semaphore = new Semaphore(this.maxConcurrentSends);
    this.sender = new MessageSender({
      outgoingManager: this.outgoing,
      semaphore,
      debug: this.debug
    });
    this.tempFiles = new TempFileManager({ debug: this.debug });
    this.tempFiles.start();
    if (config.plugins) {
      for (const plugin of config.plugins) {
        this.plugins.use(plugin);
      }
    }
    if (this.debug) {
      console.log("[SDK] Initialized");
    }
  }
  // -----------------------------------------------
  // Plugin Management
  // -----------------------------------------------
  /** Register a plugin. Late registrations are auto-initialized. */
  use(plugin) {
    this.assertAlive();
    this.plugins.use(plugin);
    return this;
  }
  // -----------------------------------------------
  // Message Queries
  // -----------------------------------------------
  /** Query messages with optional filters. */
  async getMessages(query = {}) {
    this.assertAlive();
    await this.plugins.init();
    await this.plugins.callHook("onBeforeMessageQuery", { query });
    const messages = await this.database.getMessages(query);
    await this.plugins.callHook("onAfterMessageQuery", { query, messages });
    return messages;
  }
  /** List chats with optional filters and sorting. */
  async listChats(query = {}) {
    this.assertAlive();
    await this.plugins.init();
    await this.plugins.callHook("onBeforeChatQuery", { query });
    const chats = await this.database.listChats(query);
    await this.plugins.callHook("onAfterChatQuery", { query, chats });
    return chats;
  }
  async send(toOrRequest, content) {
    this.assertAlive();
    const request = normalizeToSendRequest(toOrRequest, content);
    return this.executeSend(request);
  }
  /** Send a file attachment with optional text. */
  async sendFile(to, filePath, text) {
    return this.send({ to, text, attachments: [filePath] });
  }
  /** Send multiple file attachments with optional text. */
  async sendFiles(to, filePaths, text) {
    return this.send({ to, text, attachments: filePaths });
  }
  /** Send messages to multiple recipients with concurrency control. */
  async sendBatch(items, options = {}) {
    this.assertAlive();
    await this.plugins.init();
    const { continueOnError = true, signal } = options;
    const concurrency = options.concurrency ?? this.maxConcurrentSends;
    const semaphore = new Semaphore(Math.max(1, concurrency));
    const batchAbort = composeAbortSignals(this.abortController.signal, signal);
    const results = new Array(items.length);
    let stopError = null;
    await Promise.all(
      items.map(
        (item, index) => semaphore.run(async () => {
          if (stopError) {
            results[index] = { to: item.to, status: "skipped", error: stopError };
            return;
          }
          try {
            const result = await this.executeSend({
              to: item.to,
              text: item.text,
              attachments: item.attachments,
              signal: batchAbort
            });
            results[index] = { to: item.to, status: "sent", result };
          } catch (error) {
            const err = toError(error);
            results[index] = { to: item.to, status: "failed", error: err };
            if (!continueOnError) {
              stopError = err;
            }
          }
        }, batchAbort).catch((error) => {
          results[index] = {
            to: item.to,
            status: stopError ? "skipped" : "failed",
            error: stopError ?? toError(error)
          };
        })
      )
    );
    return {
      results,
      sent: results.filter((r) => r.status === "sent").length,
      failed: results.filter((r) => r.status === "failed").length,
      skipped: results.filter((r) => r.status === "skipped").length
    };
  }
  async executeSend(request) {
    this.assertAlive();
    await this.plugins.init();
    await this.plugins.callHook("onBeforeSend", { request });
    const signal = composeAbortSignals(this.abortController.signal, request.signal);
    const result = await this.sender.send({ ...request, signal });
    await this.plugins.callHook("onAfterSend", { request, result });
    return result;
  }
  // -----------------------------------------------
  // Chain Processing
  // -----------------------------------------------
  /** Create a fluent processing chain for a message. */
  message(message) {
    this.assertAlive();
    return new MessageChain(message, this);
  }
  // -----------------------------------------------
  // Watcher
  // -----------------------------------------------
  /** Start watching for new messages in real time. */
  async startWatching(events = {}) {
    this.assertAlive();
    if (this.watchSource) throw new Error(ERROR_WATCHER_RUNNING);
    await this.plugins.init();
    const dispatcher = new MessageDispatcher({
      events,
      sink: createPluginMessageSink(this.plugins),
      outgoingMatcher: this.outgoing,
      debug: this.debug
    });
    const watcher = new MessageWatchSource({
      database: this.database,
      databasePath: this.databasePath,
      onBatch: (messages) => dispatcher.dispatch(messages),
      onCheckpoint: events.onCheckpoint,
      onError: (error) => dispatcher.handleError(error, "watch-source"),
      debug: this.debug
    });
    try {
      await watcher.start();
      this.watchSource = watcher;
    } catch (error) {
      watcher.stop();
      throw error;
    }
  }
  /** Stop watching for new messages. */
  stopWatching() {
    this.watchSource?.stop();
    this.watchSource = null;
  }
  // -----------------------------------------------
  // Lifecycle
  // -----------------------------------------------
  /** Gracefully shut down the SDK and release all resources. */
  async close() {
    if (this.destroyed) return;
    this.destroyed = true;
    const errors = [];
    this.abortController.abort(new Error("SDK closed"));
    safeRun(errors, "watchSource", () => {
      this.watchSource?.stop();
      this.watchSource = null;
    });
    safeRun(errors, "outgoingManager", () => {
      this.outgoing.rejectAll("SDK closed");
    });
    await safeRunAsync(errors, "pluginManager", () => this.plugins.destroy());
    await safeRunAsync(errors, "tempFileManager", () => this.tempFiles.destroy());
    safeRun(errors, "database", () => {
      this.database.close();
    });
    if (errors.length > 0) {
      if (this.debug) {
        console.error("[SDK] Shutdown errors:", errors);
      }
      throw new AggregateError(
        errors.map((e) => e.error),
        `SDK shutdown failed: ${errors.map((e) => e.component).join(", ")}`
      );
    }
  }
  async [Symbol.asyncDispose]() {
    await this.close();
  }
  // -----------------------------------------------
  // Internal
  // -----------------------------------------------
  assertAlive() {
    if (this.destroyed) throw new Error(ERROR_SDK_DESTROYED);
  }
};
function validateBound(value, bound, name) {
  const resolved = value ?? bound.default;
  if (resolved < bound.min || resolved > bound.max) {
    throw ConfigError(`${name} must be between ${bound.min} and ${bound.max}, got ${resolved}`);
  }
  return resolved;
}
function normalizeToSendRequest(toOrRequest, content) {
  if (typeof toOrRequest !== "string") {
    return toOrRequest;
  }
  if (content == null) {
    throw new Error("send() requires content when called with a target string");
  }
  if (typeof content === "string") {
    return { to: toOrRequest, text: content };
  }
  return { to: toOrRequest, ...content };
}
function composeAbortSignals(...signals) {
  const active = signals.filter((s) => s != null);
  if (active.length === 0) return void 0;
  if (active.length === 1) return active[0];
  return AbortSignal.any(active);
}
function safeRun(errors, component, fn) {
  try {
    fn();
  } catch (error) {
    errors.push({ component, error: toError(error) });
  }
}
async function safeRunAsync(errors, component, fn) {
  try {
    await fn();
  } catch (error) {
    errors.push({ component, error: toError(error) });
  }
}

// src/application/message-scheduler.ts
function generateId() {
  return `sched_${crypto.randomUUID()}`;
}
function normalizeContent(content) {
  return typeof content === "string" ? { text: content } : content;
}
function validateRecurringInterval(interval) {
  if (typeof interval === "number") {
    if (!Number.isFinite(interval) || interval <= 0) {
      throw new Error("Recurrence interval must be a positive number of milliseconds");
    }
    return;
  }
  switch (interval) {
    case "hourly":
    case "daily":
    case "weekly":
    case "monthly":
      return;
    default:
      throw new Error(`Unknown recurrence interval: "${String(interval)}"`);
  }
}
function calculateNextSendTime(current, interval) {
  const next = new Date(current);
  if (typeof interval === "number") {
    if (!Number.isFinite(interval) || interval <= 0) {
      throw new Error("Recurrence interval must be a positive number of milliseconds");
    }
    next.setTime(next.getTime() + interval);
    return next;
  }
  switch (interval) {
    case "hourly":
      next.setHours(next.getHours() + 1);
      break;
    case "daily":
      next.setDate(next.getDate() + 1);
      break;
    case "weekly":
      next.setDate(next.getDate() + 7);
      break;
    case "monthly": {
      const targetDay = next.getDate();
      next.setMonth(next.getMonth() + 1);
      if (next.getDate() !== targetDay) {
        next.setDate(0);
      }
      break;
    }
    default:
      throw new Error(`Unknown recurrence interval: "${String(interval)}"`);
  }
  return next;
}
function getEffectiveSendTime(task) {
  return task.type === "recurring" ? task.nextSendAt : task.sendAt;
}
var MessageScheduler = class {
  sender;
  events;
  tickInterval;
  debug;
  tasks = /* @__PURE__ */ new Map();
  intervalHandle = null;
  destroyed = false;
  ticking = false;
  constructor(options) {
    this.sender = options.sender;
    this.events = options.events ?? {};
    this.tickInterval = options.tickInterval ?? 1e3;
    this.debug = options.debug ?? false;
  }
  // -----------------------------------------------
  // Schedule
  // -----------------------------------------------
  /** Schedule a one-time message. Returns the task id. */
  schedule(options) {
    this.assertAlive();
    const sendAt = new Date(options.sendAt);
    if (sendAt <= /* @__PURE__ */ new Date()) {
      throw new Error("sendAt must be in the future");
    }
    validateRecipient(options.to);
    const id = options.id ?? generateId();
    if (this.tasks.has(id)) {
      throw new Error(`Task "${id}" already exists`);
    }
    this.tasks.set(id, {
      id,
      type: "once",
      to: options.to,
      content: normalizeContent(options.content),
      sendAt,
      status: "pending",
      createdAt: /* @__PURE__ */ new Date()
    });
    this.log(`Scheduled task ${id} for ${sendAt.toISOString()}`);
    return id;
  }
  /** Schedule a recurring message. Returns the task id. */
  scheduleRecurring(options) {
    this.assertAlive();
    const startAt = new Date(options.startAt);
    const endAt = options.endAt ? new Date(options.endAt) : void 0;
    if (startAt <= /* @__PURE__ */ new Date()) {
      throw new Error("startAt must be in the future");
    }
    if (endAt && endAt <= startAt) {
      throw new Error("endAt must be after startAt");
    }
    validateRecipient(options.to);
    validateRecurringInterval(options.interval);
    const id = options.id ?? generateId();
    if (this.tasks.has(id)) {
      throw new Error(`Task "${id}" already exists`);
    }
    this.tasks.set(id, {
      id,
      type: "recurring",
      to: options.to,
      content: normalizeContent(options.content),
      sendAt: startAt,
      status: "pending",
      createdAt: /* @__PURE__ */ new Date(),
      interval: options.interval,
      endAt,
      sendCount: 0,
      nextSendAt: new Date(startAt)
    });
    this.log(`Scheduled recurring task ${id} starting ${startAt.toISOString()}`);
    return id;
  }
  // -----------------------------------------------
  // Management
  // -----------------------------------------------
  /** Cancel a pending task. Returns true if cancelled. */
  cancel(id) {
    const task = this.tasks.get(id);
    if (!task || task.status !== "pending") return false;
    this.tasks.delete(id);
    this.log(`Cancelled task ${id}`);
    return true;
  }
  /** Reschedule a pending one-time task. Returns true if rescheduled. */
  reschedule(id, newSendAt) {
    if (newSendAt <= /* @__PURE__ */ new Date()) {
      throw new Error("newSendAt must be in the future");
    }
    const task = this.tasks.get(id);
    if (!task || task.status !== "pending" || task.type !== "once") return false;
    this.tasks.set(id, { ...task, sendAt: new Date(newSendAt) });
    this.log(`Rescheduled task ${id} to ${newSendAt.toISOString()}`);
    return true;
  }
  /** Get a task by id. */
  get(id) {
    return this.tasks.get(id);
  }
  /** List all pending tasks sorted by send time ascending. */
  getPending() {
    const pending = [];
    for (const task of this.tasks.values()) {
      if (task.status === "pending") pending.push(task);
    }
    return pending.sort((a, b) => getEffectiveSendTime(a).getTime() - getEffectiveSendTime(b).getTime());
  }
  // -----------------------------------------------
  // Lifecycle
  // -----------------------------------------------
  /** Start the background delivery loop. */
  start() {
    this.assertAlive();
    if (this.intervalHandle) return;
    this.intervalHandle = setInterval(() => {
      this.tick().catch((err) => {
        if (this.debug) console.error("[MessageScheduler] Tick error:", err);
      });
    }, this.tickInterval);
    this.intervalHandle.unref?.();
    this.log(`Started with ${this.tickInterval}ms interval`);
  }
  /** Stop the tick loop and discard all tasks. */
  destroy() {
    if (this.destroyed) return;
    this.destroyed = true;
    if (this.intervalHandle) {
      clearInterval(this.intervalHandle);
      this.intervalHandle = null;
    }
    this.tasks.clear();
    this.log("Destroyed");
  }
  // -----------------------------------------------
  // Tick
  // -----------------------------------------------
  async tick() {
    if (this.destroyed || this.ticking) return;
    this.ticking = true;
    try {
      const now = /* @__PURE__ */ new Date();
      const due = [];
      for (const task of this.tasks.values()) {
        if (task.status === "pending" && getEffectiveSendTime(task) <= now) {
          due.push(task);
        }
      }
      for (const task of due) {
        if (!this.tasks.has(task.id)) continue;
        if (task.type === "once") {
          await this.sendOnce(task);
        } else {
          await this.sendRecurring(task);
        }
      }
    } finally {
      this.ticking = false;
    }
  }
  // -----------------------------------------------
  // Send
  // -----------------------------------------------
  async sendOnce(task) {
    try {
      this.log(`Sending task ${task.id}`);
      const result = await this.sender.send({ to: task.to, ...task.content });
      if (this.tasks.has(task.id)) {
        this.tasks.delete(task.id);
        this.emitSent({ ...task, status: "sent" }, result);
      }
    } catch (err) {
      const error = toError(err);
      if (this.tasks.has(task.id)) {
        this.tasks.delete(task.id);
        this.emitError({ ...task, status: "failed", error: error.message }, error);
      }
    }
  }
  async sendRecurring(task) {
    const nextTime = calculateNextSendTime(task.nextSendAt, task.interval);
    const isComplete = task.endAt != null && nextTime > task.endAt;
    try {
      this.log(`Sending recurring task ${task.id} (count: ${task.sendCount + 1})`);
      const result = await this.sender.send({ to: task.to, ...task.content });
      if (!this.tasks.has(task.id)) return;
      const updated = {
        ...task,
        sendCount: task.sendCount + 1,
        nextSendAt: nextTime,
        error: void 0
      };
      if (isComplete) {
        this.tasks.delete(task.id);
        this.emitSent({ ...updated, status: "sent" }, result);
        this.emitComplete(updated);
        this.log(`Recurring task ${task.id} completed`);
      } else {
        this.tasks.set(task.id, updated);
        this.emitSent(updated, result);
      }
    } catch (err) {
      const error = toError(err);
      if (!this.tasks.has(task.id)) return;
      if (isComplete) {
        this.tasks.delete(task.id);
        this.emitError({ ...task, status: "failed", error: error.message }, error);
      } else {
        this.tasks.set(task.id, { ...task, error: error.message, nextSendAt: nextTime });
        this.emitError(task, error);
      }
    }
  }
  // -----------------------------------------------
  // Events
  // -----------------------------------------------
  emitSent(task, result) {
    try {
      this.events.onSent?.(task, result);
    } catch (err) {
      if (this.debug) console.error("[MessageScheduler] onSent callback error:", err);
    }
  }
  emitError(task, error) {
    try {
      this.events.onError?.(task, error);
    } catch (err) {
      if (this.debug) console.error("[MessageScheduler] onError callback error:", err);
    }
  }
  emitComplete(task) {
    try {
      this.events.onComplete?.(task);
    } catch (err) {
      if (this.debug) console.error("[MessageScheduler] onComplete callback error:", err);
    }
  }
  // -----------------------------------------------
  // Internal
  // -----------------------------------------------
  assertAlive() {
    if (this.destroyed) throw new Error("Scheduler has been destroyed");
  }
  log(msg) {
    if (this.debug) console.log(`[MessageScheduler] ${msg}`);
  }
};

// src/application/reminder-time.ts
var WEEKDAYS = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"];
function parseDuration(duration) {
  const match = duration.toLowerCase().match(/^(\d+)\s*(second|minute|hour|day|week)s?$/);
  if (!match) {
    throw new Error(`Invalid duration format: "${duration}". Use formats like "5 minutes", "2 hours", "1 day"`);
  }
  const rawValue = match[1];
  const rawUnit = match[2];
  if (!rawValue || !rawUnit) {
    throw new Error(`Invalid duration format: "${duration}". Use formats like "5 minutes", "2 hours", "1 day"`);
  }
  const value = Number.parseInt(rawValue, 10);
  const multipliers = {
    second: 1e3,
    minute: 6e4,
    hour: 36e5,
    day: 864e5,
    week: 6048e5
  };
  const mult = multipliers[rawUnit];
  if (mult === void 0) {
    throw new Error(`Invalid duration unit: "${rawUnit}"`);
  }
  return value * mult;
}
function parseTime(timeStr) {
  const trimmed = timeStr.trim();
  const match24 = trimmed.match(/^(\d{1,2}):(\d{2})$/);
  if (match24) {
    const hours = Number.parseInt(match24[1], 10);
    const minutes = Number.parseInt(match24[2], 10);
    if (hours < 0 || hours > 23 || minutes < 0 || minutes > 59) {
      throw new Error(`Invalid time: "${timeStr}". Hours must be 0-23 and minutes 0-59 for 24-hour format`);
    }
    return { hours, minutes };
  }
  const match12 = trimmed.toLowerCase().match(/^(\d{1,2})(?::(\d{2}))?\s*(am|pm)$/);
  if (match12) {
    let hours = Number.parseInt(match12[1], 10);
    const minutes = match12[2] ? Number.parseInt(match12[2], 10) : 0;
    const period = match12[3];
    if (hours < 1 || hours > 12 || minutes < 0 || minutes > 59) {
      throw new Error(`Invalid time: "${timeStr}". Hours must be 1-12 and minutes 0-59 for 12-hour format`);
    }
    if (period === "pm" && hours !== 12) hours += 12;
    if (period === "am" && hours === 12) hours = 0;
    return { hours, minutes };
  }
  throw new Error(`Invalid time format: "${timeStr}". Use formats like "5pm", "5:30pm", "17:30"`);
}
function parseAtExpression(expression, now = /* @__PURE__ */ new Date()) {
  const parts = expression.toLowerCase().trim().split(/\s+/);
  const targetDate = new Date(now);
  let timeStr;
  let isWeekdayExpression = false;
  if (parts.length === 1) {
    const only = parts[0];
    if (!only) {
      throw new Error(`Invalid at expression: "${expression}"`);
    }
    timeStr = only;
  } else if (parts[0] === "tomorrow") {
    targetDate.setDate(targetDate.getDate() + 1);
    timeStr = parts.slice(1).join(" ");
  } else if (WEEKDAYS.includes(parts[0])) {
    isWeekdayExpression = true;
    const targetDay = WEEKDAYS.indexOf(parts[0]);
    const currentDay = now.getDay();
    const rawDiff = targetDay - currentDay;
    const daysUntil = rawDiff < 0 ? rawDiff + 7 : rawDiff;
    targetDate.setDate(targetDate.getDate() + daysUntil);
    timeStr = parts.slice(1).join(" ");
  } else {
    timeStr = expression;
  }
  const { hours, minutes } = parseTime(timeStr);
  targetDate.setHours(hours, minutes, 0, 0);
  if (targetDate <= now) {
    targetDate.setDate(targetDate.getDate() + (isWeekdayExpression ? 7 : 1));
  }
  return targetDate;
}

// src/application/reminders.ts
var Reminders = class {
  scheduler;
  entries = /* @__PURE__ */ new Map();
  constructor(sender, events) {
    this.scheduler = new MessageScheduler({
      sender,
      tickInterval: 1e3,
      events: {
        onSent: (task, result) => {
          if (task.status !== "pending") this.entries.delete(task.id);
          events?.onSent?.(task, result);
        },
        onError: (task, error) => {
          if (task.status !== "pending") this.entries.delete(task.id);
          events?.onError?.(task, error);
        },
        onComplete: (task) => {
          this.entries.delete(task.id);
          events?.onComplete?.(task);
        }
      }
    });
    this.scheduler.start();
  }
  // -----------------------------------------------
  // Schedule
  // -----------------------------------------------
  /** Schedule a reminder relative to now. e.g. "5 minutes", "2 hours" */
  in(duration, to, message, options) {
    const ms = parseDuration(duration);
    const sendAt = new Date(Date.now() + ms);
    return this.scheduleReminder(to, message, sendAt, options);
  }
  /** Schedule at a natural language time. e.g. "5pm", "tomorrow 9am", "friday 17:30" */
  at(timeExpression, to, message, options) {
    const sendAt = parseAtExpression(timeExpression);
    return this.scheduleReminder(to, message, sendAt, options);
  }
  /** Schedule at an exact Date. */
  exact(date, to, message, options) {
    return this.scheduleReminder(to, message, date, options);
  }
  // -----------------------------------------------
  // Management
  // -----------------------------------------------
  cancel(id) {
    const cancelled = this.scheduler.cancel(id);
    if (cancelled) this.entries.delete(id);
    return cancelled;
  }
  list() {
    return Array.from(this.entries.values()).filter((r) => this.scheduler.get(r.id)?.status === "pending").sort((a, b) => a.scheduledFor.getTime() - b.scheduledFor.getTime());
  }
  count() {
    let count = 0;
    for (const r of this.entries.values()) {
      if (this.scheduler.get(r.id)?.status === "pending") count++;
    }
    return count;
  }
  // -----------------------------------------------
  // Lifecycle
  // -----------------------------------------------
  destroy() {
    this.scheduler.destroy();
    this.entries.clear();
  }
  // -----------------------------------------------
  // Internal
  // -----------------------------------------------
  scheduleReminder(to, message, sendAt, options) {
    const emoji = options?.emoji ?? "\u23F0";
    const formattedMessage = `${emoji} Reminder: ${message}`;
    const id = this.scheduler.schedule({
      id: options?.id,
      to,
      content: formattedMessage,
      sendAt
    });
    this.entries.set(id, {
      id,
      to,
      message,
      scheduledFor: sendAt,
      createdAt: /* @__PURE__ */ new Date()
    });
    return id;
  }
};

// src/infra/plugin/logger.ts
var LEVELS = { debug: 0, info: 1, warn: 2, error: 3 };
var COLORS = {
  debug: "\x1B[36m",
  info: "\x1B[32m",
  warn: "\x1B[33m",
  error: "\x1B[31m"
};
var RESET = "\x1B[0m";
var loggerPlugin = (options = {}) => {
  const { level = "info", colors = true, timestamps = false, logSend = true, logNewMessage = false } = options;
  const log = (logLevel, message, data) => {
    if (LEVELS[logLevel] < LEVELS[level]) return;
    const time = timestamps ? (/* @__PURE__ */ new Date()).toISOString() : "";
    const tag = logLevel.toUpperCase().padEnd(5);
    const color = colors ? COLORS[logLevel] : "";
    const reset = colors ? RESET : "";
    const prefix = time ? `${time} ${color}[${tag}]${reset}` : `${color}[${tag}]${reset}`;
    const output = `${prefix} ${message}`;
    if (logLevel === "error") {
      console.error(output);
    } else if (logLevel === "warn") {
      console.warn(output);
    } else {
      console.log(output);
    }
  };
  return {
    name: "logger",
    version: "1.0.0",
    description: "Built-in logger plugin",
    onInit: () => {
      log("info", "SDK initialized");
    },
    onBeforeSend: ({ request }) => {
      if (!logSend) return;
      const preview = request.text?.substring(0, 30) || "(no text)";
      const attachCount = request.attachments?.length || 0;
      const attachInfo = attachCount ? ` + ${attachCount} attachment(s)` : "";
      log("info", `[SEND] Sending to ${request.to}: ${preview}${attachInfo}`);
    },
    onAfterSend: ({ request }) => {
      if (logSend) {
        log("info", `[OK] Sent successfully -> ${request.to}`);
      }
    },
    onNewMessage: ({ message }) => {
      if (!logNewMessage) return;
      const preview = message.text?.substring(0, 40) || "(no text)";
      const attachCount = message.attachments.length;
      const attachInfo = attachCount ? ` [${attachCount}]` : "";
      const source = message.participant ?? (message.isFromMe ? "me" : "unknown");
      log("info", `[MSG] New message from ${source}: ${preview}${attachInfo}`);
    },
    onError: ({ error, context }) => {
      log("error", `[ERROR] ${context || "Error"}: ${error.message}`);
    },
    onDestroy: () => {
      log("info", "[CLOSE] SDK destroyed");
    }
  };
};
var IMAGE_EXTENSIONS = /* @__PURE__ */ new Set(["jpg", "jpeg", "png", "gif", "heic", "heif", "webp", "bmp", "tiff", "svg"]);
var VIDEO_EXTENSIONS = /* @__PURE__ */ new Set(["mp4", "mov", "avi", "mkv", "m4v", "wmv", "flv", "webm"]);
var AUDIO_EXTENSIONS = /* @__PURE__ */ new Set(["mp3", "m4a", "wav", "aac", "flac", "ogg", "wma"]);
function requireLocalPath(attachment) {
  if (!attachment.localPath) {
    throw new Error("Attachment does not have a local file path");
  }
  return attachment.localPath;
}
async function attachmentExists(attachment) {
  if (!attachment.localPath) return false;
  try {
    await promises.access(attachment.localPath);
    return true;
  } catch {
    return false;
  }
}
async function copyAttachmentFile(attachment, destPath) {
  const localPath = requireLocalPath(attachment);
  const destDir = path.dirname(destPath);
  await promises.mkdir(destDir, { recursive: true });
  await promises.copyFile(localPath, destPath);
}
async function readAttachmentBytes(attachment) {
  return await promises.readFile(requireLocalPath(attachment));
}
async function getAttachmentSize(attachment) {
  const info = await getAttachmentFileInfo(attachment);
  return info?.size ?? 0;
}
async function getAttachmentFileInfo(attachment) {
  if (!attachment.localPath) return null;
  try {
    const stats = await promises.stat(attachment.localPath);
    return {
      localPath: attachment.localPath,
      size: stats.size,
      createdAt: stats.birthtime,
      modifiedAt: stats.mtime
    };
  } catch {
    return null;
  }
}
function getAttachmentExtension(attachment) {
  const ext = path.extname(attachment.fileName ?? attachment.localPath ?? "");
  return ext ? ext.slice(1).toLowerCase() : "";
}
function isImageAttachment(attachment) {
  return IMAGE_EXTENSIONS.has(getAttachmentExtension(attachment));
}
function isVideoAttachment(attachment) {
  return VIDEO_EXTENSIONS.has(getAttachmentExtension(attachment));
}
function isAudioAttachment(attachment) {
  return AUDIO_EXTENSIONS.has(getAttachmentExtension(attachment));
}

exports.BOUNDS = BOUNDS;
exports.ChatId = ChatId;
exports.ConfigError = ConfigError;
exports.DatabaseError = DatabaseError;
exports.IMessageError = IMessageError;
exports.IMessageSDK = IMessageSDK;
exports.LIMITS = SEND_LIMITS;
exports.MessageChain = MessageChain;
exports.MessageDispatcher = MessageDispatcher;
exports.MessageScheduler = MessageScheduler;
exports.PlatformError = PlatformError;
exports.Reminders = Reminders;
exports.SendError = SendError;
exports.attachmentExists = attachmentExists;
exports.copyAttachmentFile = copyAttachmentFile;
exports.definePlugin = definePlugin;
exports.getAttachmentExtension = getAttachmentExtension;
exports.getAttachmentFileInfo = getAttachmentFileInfo;
exports.getAttachmentSize = getAttachmentSize;
exports.getDefaultDatabasePath = getDefaultDatabasePath;
exports.isAudioAttachment = isAudioAttachment;
exports.isImageAttachment = isImageAttachment;
exports.isURL = isURL;
exports.isVideoAttachment = isVideoAttachment;
exports.loggerPlugin = loggerPlugin;
exports.parseAtExpression = parseAtExpression;
exports.parseDuration = parseDuration;
exports.readAttachmentBytes = readAttachmentBytes;
exports.requireMacOS = requireMacOS;
exports.resolveTarget = resolveTarget;
exports.toError = toError;
exports.toErrorMessage = toErrorMessage;
exports.validateRecipient = validateRecipient;
