// spectrum-ts/dist/providers/whatsapp-business/index.js
import {
  asAttachment,
  asCustom,
  stream
} from "../../chunk-5XW4CAWS.js";
import {
  asText,
  definePlatform
} from "../../chunk-XEEDIGVK.js";

// src/providers/whatsapp-business/index.ts
import {
  createClient
} from "@photon-ai/whatsapp-business";

// src/providers/whatsapp-business/messages.ts
var toMessage = (client, msg) => ({
  id: msg.id,
  content: mapContent(client, msg.content),
  sender: { id: msg.from },
  space: { id: msg.from },
  timestamp: msg.timestamp
});
var mapContent = (client, content) => {
  switch (content.type) {
    case "text":
      return asText(content.body);
    case "image":
    case "video":
    case "audio":
    case "document":
      return lazyMedia(client, content.media);
    case "sticker":
      return asCustom({ whatsapp_type: "sticker", ...content.sticker });
    case "location":
      return asCustom({ whatsapp_type: "location", ...content.location });
    case "contacts":
      return asCustom({
        whatsapp_type: "contacts",
        contacts: content.contacts
      });
    case "reaction":
      return asCustom({ whatsapp_type: "reaction", ...content.reaction });
    case "interactive":
      return asCustom({ whatsapp_type: "interactive", ...content.interactive });
    case "button":
      return asCustom({ whatsapp_type: "button", ...content.button });
    case "order":
      return asCustom({ whatsapp_type: "order", ...content.order });
    case "system":
      return asCustom({ whatsapp_type: "system", ...content.system });
    default:
      return asCustom({ whatsapp_type: "unknown" });
  }
};
var fetchMedia = async (client, mediaId) => {
  const { url } = await client.media.getUrl(mediaId);
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Media download failed: ${response.status}`);
  }
  return response;
};
var lazyMedia = (client, media) => asAttachment({
  name: media.filename ?? `media-${media.id}`,
  mimeType: media.mimeType,
  read: async () => Buffer.from(await (await fetchMedia(client, media.id)).arrayBuffer()),
  stream: async () => {
    const response = await fetchMedia(client, media.id);
    if (!response.body) {
      throw new Error("Media response missing body");
    }
    return response.body;
  }
});
var mimeToMediaType = (mimeType) => {
  if (mimeType.startsWith("image/")) {
    return "image";
  }
  if (mimeType.startsWith("video/")) {
    return "video";
  }
  if (mimeType.startsWith("audio/")) {
    return "audio";
  }
  return "document";
};
var messages = (client) => {
  const eventStream = client.events.subscribe().filter(
    (e) => e.type === "message"
  );
  return stream((emit, end) => {
    (async () => {
      try {
        for await (const event of eventStream) {
          emit(toMessage(client, event.message));
        }
        end();
      } catch (e) {
        end(e);
      }
    })();
    return () => eventStream.close();
  });
};
var send = async (client, spaceId, content) => {
  switch (content.type) {
    case "text":
      await client.messages.send({ to: spaceId, text: content.text });
      break;
    case "attachment": {
      const { mediaId } = await client.media.upload({
        file: await content.read(),
        mimeType: content.mimeType,
        filename: content.name
      });
      const mediaType = mimeToMediaType(content.mimeType);
      const mediaPayload = mediaType === "document" ? { id: mediaId, filename: content.name } : { id: mediaId };
      await client.messages.send({
        to: spaceId,
        [mediaType]: mediaPayload
      });
      break;
    }
    default:
      break;
  }
};
var reactToMessage = async (client, spaceId, messageId, reaction) => {
  await client.messages.send({
    to: spaceId,
    reaction: { messageId, emoji: reaction }
  });
};
var replyToMessage = async (client, spaceId, messageId, content) => {
  switch (content.type) {
    case "text":
      await client.messages.send({
        to: spaceId,
        replyTo: messageId,
        text: content.text
      });
      break;
    case "attachment": {
      const { mediaId } = await client.media.upload({
        file: await content.read(),
        mimeType: content.mimeType,
        filename: content.name
      });
      const mediaType = mimeToMediaType(content.mimeType);
      const mediaPayload = mediaType === "document" ? { id: mediaId, filename: content.name } : { id: mediaId };
      await client.messages.send({
        to: spaceId,
        replyTo: messageId,
        [mediaType]: mediaPayload
      });
      break;
    }
    default:
      break;
  }
};

// src/providers/whatsapp-business/types.ts
import z from "zod";
var configSchema = z.object({
  accessToken: z.string().min(1),
  phoneNumberId: z.string().min(1),
  appSecret: z.string().optional()
});
var userSchema = z.object({});
var spaceSchema = z.object({
  id: z.string()
});

// src/providers/whatsapp-business/index.ts
var whatsappBusiness = definePlatform("WhatsApp Business", {
  config: configSchema,
  user: {
    resolve: async ({ input }) => ({ id: input.userID })
  },
  space: {
    schema: spaceSchema,
    resolve: async ({ input }) => {
      if (input.users.length === 0) {
        throw new Error("WhatsApp space creation requires at least one user");
      }
      if (input.users.length > 1) {
        throw new Error(
          "WhatsApp Business API only supports 1:1 conversations"
        );
      }
      const user = input.users[0];
      if (!user) {
        throw new Error("WhatsApp space creation requires a user");
      }
      return { id: user.id };
    }
  },
  lifecycle: {
    createClient: async ({ config }) => {
      return createClient({
        accessToken: config.accessToken,
        phoneNumberId: config.phoneNumberId,
        appSecret: config.appSecret ?? ""
      });
    },
    destroyClient: async ({ client }) => {
      await client.close();
    }
  },
  events: {
    messages: ({ client }) => messages(client)
  },
  actions: {
    send: async ({ space, content, client }) => {
      await send(client, space.id, content);
    },
    reactToMessage: async ({ space, messageId, reaction, client }) => {
      await reactToMessage(
        client,
        space.id,
        messageId,
        reaction
      );
    },
    replyToMessage: async ({ space, messageId, content, client }) => {
      await replyToMessage(
        client,
        space.id,
        messageId,
        content
      );
    }
  }
});
export {
  whatsappBusiness
};
