// spectrum-ts/dist/providers/imessage/index.js
import {
  cloud
} from "../../chunk-HXM64ENV.js";
import {
  asAttachment,
  asCustom,
  mergeStreams,
  stream
} from "../../chunk-5XW4CAWS.js";
import {
  asText,
  definePlatform
} from "../../chunk-XEEDIGVK.js";

// src/providers/imessage/index.ts
import { createClient as createClient2, directChat } from "@photon-ai/advanced-imessage";
import { IMessageSDK as IMessageSDK2 } from "@photon-ai/imessage-kit";

// src/providers/imessage/auth.ts
import {
  createClient
} from "@photon-ai/advanced-imessage";
var RENEWAL_RATIO = 0.8;
var EXPIRY_BUFFER_MS = 3e4;
var RETRY_DELAY_MS = 3e4;
var cloudAuthState = /* @__PURE__ */ new WeakMap();
async function createCloudClients(projectId, projectSecret) {
  let tokenData = await cloud.issueImessageTokens(projectId, projectSecret);
  let tokenExpiresAt = Date.now() + tokenData.expiresIn * 1e3;
  let disposed = false;
  let renewalTimer;
  const scheduleRenewal = () => {
    if (disposed) {
      return;
    }
    const ttlMs = tokenData.expiresIn * 1e3;
    const renewInMs = Math.max(ttlMs * RENEWAL_RATIO, 5e3);
    renewalTimer = setTimeout(async () => {
      try {
        tokenData = await cloud.issueImessageTokens(projectId, projectSecret);
        tokenExpiresAt = Date.now() + tokenData.expiresIn * 1e3;
        scheduleRenewal();
      } catch {
        renewalTimer = setTimeout(() => scheduleRenewal(), RETRY_DELAY_MS);
        renewalTimer?.unref?.();
      }
    }, renewInMs);
    renewalTimer?.unref?.();
  };
  scheduleRenewal();
  const refreshIfNeeded = async () => {
    if (Date.now() < tokenExpiresAt - EXPIRY_BUFFER_MS) {
      return;
    }
    tokenData = await cloud.issueImessageTokens(projectId, projectSecret);
    tokenExpiresAt = Date.now() + tokenData.expiresIn * 1e3;
    scheduleRenewal();
  };
  const buildClients = () => {
    if (tokenData.type === "shared") {
      const address = process.env.SPECTRUM_IMESSAGE_ADDRESS ?? "imessage.spectrum.photon.codes:443";
      return [
        createClient({
          address,
          tls: true,
          token: async () => {
            await refreshIfNeeded();
            return tokenData.token;
          }
        })
      ];
    }
    return Object.entries(tokenData.auth).map(
      ([instanceId, token]) => createClient({
        address: `${instanceId}.imsg.photon.codes:443`,
        tls: true,
        token: async () => {
          await refreshIfNeeded();
          const data = tokenData;
          return data.auth[instanceId] ?? token;
        }
      })
    );
  };
  const clients = buildClients();
  cloudAuthState.set(clients, {
    dispose: () => {
      disposed = true;
      if (renewalTimer !== void 0) {
        clearTimeout(renewalTimer);
        renewalTimer = void 0;
      }
    }
  });
  return clients;
}
async function disposeCloudAuth(clients) {
  const auth = cloudAuthState.get(clients);
  if (auth) {
    auth.dispose();
    cloudAuthState.delete(clients);
  }
}

// src/providers/imessage/local.ts
import { createReadStream } from "fs";
import { unlink, writeFile } from "fs/promises";
import { tmpdir } from "os";
import { join } from "path";
import { Readable } from "stream";
import {
  readAttachmentBytes
} from "@photon-ai/imessage-kit";
var DEFAULT_ATTACHMENT_NAME = "attachment";
var toSpace = (message) => ({
  id: message.chatId,
  type: message.chatKind === "group" ? "group" : "dm"
});
var toMessages = (message) => {
  const base = {
    sender: { id: message.participant ?? "" },
    space: toSpace(message),
    timestamp: message.createdAt
  };
  if (message.attachments.length > 0) {
    return message.attachments.map((att) => {
      const { localPath } = att;
      return {
        ...base,
        id: `${message.id}:${att.id}`,
        content: asAttachment({
          name: att.fileName ?? DEFAULT_ATTACHMENT_NAME,
          mimeType: att.mimeType,
          size: att.sizeBytes,
          read: () => readAttachmentBytes(att),
          stream: localPath ? async () => Readable.toWeb(
            createReadStream(localPath)
          ) : void 0
        })
      };
    });
  }
  return [
    {
      ...base,
      id: message.id,
      content: { type: "text", text: message.text ?? "" }
    }
  ];
};
var messages = (client) => stream((emit, end) => {
  client.startWatching({
    onMessage: (message) => {
      try {
        for (const m of toMessages(message)) {
          emit(m);
        }
      } catch (error) {
        end(error);
      }
    }
  });
  return () => client.stopWatching();
});
var send = async (client, spaceId, content) => {
  switch (content.type) {
    case "text":
      await client.send(spaceId, content.text);
      break;
    case "attachment": {
      const tmp = join(tmpdir(), `spectrum-${Date.now()}-${content.name}`);
      await writeFile(tmp, await content.read());
      try {
        await client.send(spaceId, { attachments: [tmp] });
      } finally {
        await unlink(tmp).catch(() => {
        });
      }
      break;
    }
    default:
      break;
  }
};

// src/providers/imessage/remote.ts
import {
  chatGuid,
  messageGuid,
  Reaction
} from "@photon-ai/advanced-imessage";
var TAPBACK_NAMES = new Set(
  Object.values(Reaction).filter((r) => r !== "emoji" && r !== "sticker")
);
var baseMessage = (event) => ({
  sender: { id: event.message.sender?.address ?? "" },
  space: {
    id: event.chatGuid,
    type: event.chatGuid.includes(";+;") ? "group" : "dm"
  },
  timestamp: event.timestamp
});
var toMessages2 = (client, event) => {
  const base = baseMessage(event);
  const messageGuidStr = event.message.guid;
  if (event.message.attachments.length > 0) {
    return event.message.attachments.map((info) => ({
      ...base,
      id: `${messageGuidStr}:${info.guid}`,
      content: asAttachment({
        name: info.fileName,
        mimeType: info.mimeType,
        size: info.totalBytes,
        read: async () => Buffer.from(await client.attachments.downloadBuffer(info.guid)),
        stream: async () => client.attachments.download(info.guid).stream
      })
    }));
  }
  const text = event.message.text;
  return [
    {
      ...base,
      id: messageGuidStr,
      content: text ? asText(text) : asCustom(event.message)
    }
  ];
};
var clientStream = (client) => {
  const sub = client.messages.subscribe("message.received");
  return stream((emit, end) => {
    (async () => {
      try {
        for await (const event of sub) {
          for (const message of toMessages2(client, event)) {
            emit(message);
          }
        }
        end();
      } catch (e) {
        end(e);
      }
    })();
    return () => sub.close();
  });
};
var messages2 = (clients) => mergeStreams(clients.map(clientStream));
var startTyping = async (clients, spaceId) => {
  const remote = clients[0];
  if (!remote) {
    return;
  }
  await remote.chats.startTyping(chatGuid(spaceId));
};
var stopTyping = async (clients, spaceId) => {
  const remote = clients[0];
  if (!remote) {
    return;
  }
  await remote.chats.stopTyping(chatGuid(spaceId));
};
var send2 = async (clients, spaceId, content) => {
  const remote = clients[0];
  if (!remote) {
    return;
  }
  switch (content.type) {
    case "text":
      await remote.messages.send(chatGuid(spaceId), content.text);
      break;
    case "attachment": {
      const attachment = await remote.attachments.upload({
        data: await content.read(),
        fileName: content.name,
        mimeType: content.mimeType
      });
      await remote.messages.send(chatGuid(spaceId), "", {
        attachment: attachment.guid
      });
      break;
    }
    default:
      throw new Error(`Unsupported iMessage content type: ${content.type}`);
  }
};
var replyToMessage = async (clients, spaceId, msgId, content) => {
  const remote = clients[0];
  if (!remote) {
    return;
  }
  const chat = chatGuid(spaceId);
  const replyTo = messageGuid(msgId);
  switch (content.type) {
    case "text":
      await remote.messages.send(chat, content.text, { replyTo });
      break;
    case "attachment": {
      const attachment = await remote.attachments.upload({
        data: await content.read(),
        fileName: content.name,
        mimeType: content.mimeType
      });
      await remote.messages.send(chat, "", {
        attachment: attachment.guid,
        replyTo
      });
      break;
    }
    default:
      throw new Error(`Unsupported iMessage content type: ${content.type}`);
  }
};
var reactToMessage = async (clients, spaceId, msgId, reaction) => {
  const remote = clients[0];
  if (!remote) {
    return;
  }
  const chat = chatGuid(spaceId);
  const msg = messageGuid(msgId);
  if (TAPBACK_NAMES.has(reaction)) {
    await remote.messages.react(chat, msg, reaction);
  } else {
    await remote.messages.reactEmoji(chat, msg, reaction);
  }
};

// src/providers/imessage/types.ts
import { IMessageSDK } from "@photon-ai/imessage-kit";
import z from "zod";
var isLocal = (client) => client instanceof IMessageSDK;
var clientEntry = z.object({ address: z.string(), token: z.string() });
var configSchema = z.union([
  z.object({ local: z.literal(true) }),
  z.object({
    local: z.literal(false).optional().default(false),
    clients: clientEntry.or(z.array(clientEntry)).optional()
  })
]);
var userSchema = z.object({});
var spaceSchema = z.object({
  id: z.string(),
  type: z.enum(["dm", "group"])
});

// src/providers/imessage/index.ts
var imessage = definePlatform("iMessage", {
  config: configSchema,
  static: {
    tapbacks: {
      love: "love",
      like: "like",
      dislike: "dislike",
      laugh: "laugh",
      emphasize: "emphasize",
      question: "question"
    }
  },
  user: {
    resolve: async ({ input }) => ({ id: input.userID })
  },
  space: {
    schema: spaceSchema,
    resolve: async ({ input, client }) => {
      if (isLocal(client)) {
        throw new Error(
          "Space creation is not supported in local mode. Local mode only supports replying to messages."
        );
      }
      if (input.users.length === 0) {
        throw new Error("iMessage space creation requires at least one user");
      }
      const addresses = input.users.map((u) => u.id);
      if (input.users.length === 1) {
        return {
          id: directChat(addresses[0] ?? ""),
          type: "dm"
        };
      }
      const remote = client[0];
      if (!remote) {
        throw new Error("No remote iMessage client available");
      }
      const { chat } = await remote.chats.create(addresses);
      return { id: chat.guid, type: "group" };
    }
  },
  lifecycle: {
    createClient: async ({
      config,
      projectId,
      projectSecret
    }) => {
      if (config.local) {
        return new IMessageSDK2();
      }
      if (config.clients) {
        const entries = Array.isArray(config.clients) ? config.clients : [config.clients];
        return entries.map(
          (e) => createClient2({ address: e.address, tls: true, token: e.token })
        );
      }
      if (!(projectId && projectSecret)) {
        throw new Error(
          "iMessage requires projectId and projectSecret. Either pass credentials to Spectrum(), use local mode: imessage.config({ local: true }), or provide explicit client config: imessage.config({ clients: [...] })"
        );
      }
      return await createCloudClients(projectId, projectSecret);
    },
    destroyClient: async ({ client }) => {
      if (isLocal(client)) {
        await client.close();
        return;
      }
      await disposeCloudAuth(client);
      await Promise.all(client.map((c) => c.close()));
    }
  },
  events: {
    messages: ({ client }) => isLocal(client) ? messages(client) : messages2(client)
  },
  actions: {
    send: async ({ space, content, client }) => {
      if (isLocal(client)) {
        await send(client, space.id, content);
      } else {
        await send2(client, space.id, content);
      }
    },
    startTyping: async ({ space, client }) => {
      if (isLocal(client)) {
        return;
      }
      await startTyping(client, space.id);
    },
    stopTyping: async ({ space, client }) => {
      if (isLocal(client)) {
        return;
      }
      await stopTyping(client, space.id);
    },
    reactToMessage: async ({ space, messageId, reaction, client }) => {
      if (isLocal(client)) {
        return;
      }
      await reactToMessage(client, space.id, messageId, reaction);
    },
    replyToMessage: async ({ space, messageId, content, client }) => {
      if (isLocal(client)) {
        return;
      }
      await replyToMessage(client, space.id, messageId, content);
    }
  }
});
export {
  imessage
};
