// spectrum-ts/dist/providers/terminal/index.js
import {
  definePlatform
} from "../../chunk-XEEDIGVK.js";

// src/providers/terminal/index.ts
import { createInterface } from "readline";
import z from "zod";
var terminal = definePlatform("terminal", {
  config: z.object({}),
  user: {
    resolve: async ({ input }) => ({
      id: input.userID
    })
  },
  space: {
    resolve: async () => ({
      id: "terminal"
    })
  },
  lifecycle: {
    createClient: async () => {
      const client = createInterface({
        input: process.stdin,
        output: process.stdout
      });
      client.on("SIGINT", () => {
        client.close();
        process.kill(process.pid, "SIGINT");
      });
      return client;
    },
    destroyClient: async ({ client }) => {
      client.close();
      process.stdin.unref();
    }
  },
  events: {
    async *messages({ client }) {
      for await (const line of client) {
        yield {
          id: crypto.randomUUID(),
          content: { type: "text", text: line },
          sender: { id: "terminal-user" },
          space: { id: "terminal" },
          timestamp: /* @__PURE__ */ new Date()
        };
      }
    }
  },
  actions: {
    send: async ({ content }) => {
      if (content.type === "text") {
        console.log(content.text);
      }
    }
  }
});
export {
  terminal
};
